<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>5&nbsp; Data Manipulation</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./12_quality_control.html" rel="next">
<link href="./04_containers.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

<script src="site_libs/kePrint-0.0.1/kePrint.js"></script>
<link href="site_libs/lightable-0.0.1/lightable.css" rel="stylesheet">


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./10_manipulation.html"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data Manipulation</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Orchestrating Microbiome Analysis with Bioconductor</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="" rel="" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-twitter"></i></a>
    <a href="https://microbiome.github.io/mia/" rel="" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-facebook"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Orchestrating Microbiome Analysis with Bioconductor</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10_manipulation.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11_taxonomic_information.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">global knitr options</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./24_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./30_differential_abundance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Developers</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Sessioninfo</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#tidying-and-subsetting" id="toc-tidying-and-subsetting" class="nav-link active" data-scroll-target="#tidying-and-subsetting"><span class="header-section-number">5.1</span> Tidying and subsetting</a>
  <ul class="collapse">
  <li><a href="#tidy-data" id="toc-tidy-data" class="nav-link" data-scroll-target="#tidy-data"><span class="header-section-number">5.1.1</span> Tidy data</a></li>
  <li><a href="#subsetting" id="toc-subsetting" class="nav-link" data-scroll-target="#subsetting"><span class="header-section-number">5.1.2</span> Subsetting</a></li>
  <li><a href="#splitting" id="toc-splitting" class="nav-link" data-scroll-target="#splitting"><span class="header-section-number">5.1.3</span> Splitting</a></li>
  </ul></li>
  <li><a href="#add-or-modify-data" id="toc-add-or-modify-data" class="nav-link" data-scroll-target="#add-or-modify-data"><span class="header-section-number">5.2</span> Add or modify data</a></li>
  <li><a href="#merge-data" id="toc-merge-data" class="nav-link" data-scroll-target="#merge-data"><span class="header-section-number">5.3</span> Merge data</a>
  <ul class="collapse">
  <li><a href="#additional-functions" id="toc-additional-functions" class="nav-link" data-scroll-target="#additional-functions"><span class="header-section-number">5.3.1</span> Additional functions</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="datamanipulation" class="quarto-section-identifier"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<section id="tidying-and-subsetting" class="level2" data-number="5.1">
<h2 data-number="5.1" class="anchored" data-anchor-id="tidying-and-subsetting"><span class="header-section-number">5.1</span> Tidying and subsetting</h2>
<section id="tidy-data" class="level3" data-number="5.1.1">
<h3 data-number="5.1.1" class="anchored" data-anchor-id="tidy-data"><span class="header-section-number">5.1.1</span> Tidy data</h3>
<p>For several custom analysis and visualization packages, such as those from <code>tidyverse</code>, the <code>SE</code> data can be converted to a long data.frame format with <code>meltAssay</code>.</p>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-2_649284163fb22911c965e8331aab81f1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(GlobalPatterns, <span class="at">package=</span><span class="st">"mia"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> GlobalPatterns</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">MARGIN =</span> <span class="st">"samples"</span>, <span class="at">method=</span><span class="st">"relabundance"</span>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>molten_tse <span class="ot">&lt;-</span> mia<span class="sc">::</span><span class="fu">meltAssay</span>(tse,</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>                        <span class="at">add_row_data =</span> <span class="cn">TRUE</span>,</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>                        <span class="at">add_col_data =</span> <span class="cn">TRUE</span>,</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>                        <span class="at">assay.type =</span> <span class="st">"relabundance"</span>)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a>molten_tse</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 499,616 × 17
   FeatureID SampleID relabundance Kingdom Phylum       Class Order Family Genus
   &lt;fct&gt;     &lt;fct&gt;           &lt;dbl&gt; &lt;chr&gt;   &lt;chr&gt;        &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;  &lt;chr&gt;
 1 549322    CL3                 0 Archaea Crenarchaeo… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
 2 549322    CC1                 0 Archaea Crenarchaeo… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
 3 549322    SV1                 0 Archaea Crenarchaeo… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
 4 549322    M31Fcsw             0 Archaea Crenarchaeo… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
 5 549322    M11Fcsw             0 Archaea Crenarchaeo… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
 6 549322    M31Plmr             0 Archaea Crenarchaeo… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
 7 549322    M11Plmr             0 Archaea Crenarchaeo… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
 8 549322    F21Plmr             0 Archaea Crenarchaeo… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
 9 549322    M31Tong             0 Archaea Crenarchaeo… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
10 549322    M11Tong             0 Archaea Crenarchaeo… Ther… &lt;NA&gt;  &lt;NA&gt;   &lt;NA&gt; 
# ℹ 499,606 more rows
# ℹ 8 more variables: Species &lt;chr&gt;, X.SampleID &lt;fct&gt;, Primer &lt;fct&gt;,
#   Final_Barcode &lt;fct&gt;, Barcode_truncated_plus_T &lt;fct&gt;,
#   Barcode_full_length &lt;fct&gt;, SampleType &lt;fct&gt;, Description &lt;fct&gt;</code></pre>
</div>
</div>
</section>
<section id="subsetting" class="level3" data-number="5.1.2">
<h3 data-number="5.1.2" class="anchored" data-anchor-id="subsetting"><span class="header-section-number">5.1.2</span> Subsetting</h3>
<p><strong>Subsetting</strong> data helps to draw the focus of analysis on particular sets of samples and / or features. When dealing with large datasets, the subset of interest can be extracted and investigated separately. This might improve performance and reduce the computational load.</p>
<p>Load:</p>
<ul>
<li>mia</li>
<li>dplyr</li>
<li>knitr</li>
<li>data <code>GlobalPatterns</code></li>
</ul>
<p>Let us store <code>GlobalPatterns</code> into <code>tse</code> and check its original number of features (rows) and samples (columns). <strong>Note</strong>: when subsetting by sample, expect the number of columns to decrease; when subsetting by feature, expect the number of rows to decrease.</p>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-4_6ab02e4420ba99b54af64c266beed78e">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Store data into se and check dimensions</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"GlobalPatterns"</span>, <span class="at">package=</span><span class="st">"mia"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> GlobalPatterns</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Show dimensions (features x samples)</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse) </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 19216    26</code></pre>
</div>
</div>
<section id="subset-by-sample-column-wise" class="level4" data-number="5.1.2.1">
<h4 data-number="5.1.2.1" class="anchored" data-anchor-id="subset-by-sample-column-wise"><span class="header-section-number">5.1.2.1</span> Subset by sample (column-wise)</h4>
<p>For the sake of demonstration, here we will extract a subset containing only the samples of human origin (feces, skin or tongue), stored as <code>SampleType</code> within <code>colData(tse)</code> and also in <code>tse</code>.</p>
<p>First, we would like to see all the possible values that <code>SampleType</code> can take on and how frequent those are:</p>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-5_10a594c6033f3bdb59ed5b3168a6d466">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect possible values for SampleType</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(tse<span class="sc">$</span>SampleType)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] Soil               Feces              Skin               Tongue            
[5] Freshwater         Freshwater (creek) Ocean              Sediment (estuary)
[9] Mock              
9 Levels: Feces Freshwater Freshwater (creek) Mock ... Tongue</code></pre>
</div>
</div>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-6_7dab238f637f5d9a96547935a1e2854e">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Show the frequency of each value</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>tse<span class="sc">$</span>SampleType <span class="sc">%&gt;%</span> <span class="fu">table</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-7_aba08d80b435e08ae6f9b2e79f020f68">
<div class="cell-output-display">
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table table-striped table-sm small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">.</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Feces</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Freshwater</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Freshwater (creek)</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Mock</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Ocean</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Sediment (estuary)</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Skin</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Soil</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Tongue</td>
<td style="text-align: right;">2</td>
</tr>
</tbody>
</table>
</div>

</div>
</div>
<p><strong>Note</strong>: after subsetting, expect the number of columns to equal the sum of the frequencies of the samples that you are interested in. For instance, <code>ncols = Feces + Skin + Tongue = 4 + 3 + 2 = 9</code>.</p>
<p>Next, we <em>logical index</em> across the columns of <code>tse</code> (make sure to leave the first index empty to select all rows) and filter for the samples of human origin. For this, we use the information on the samples from the meta data <code>colData(tse)</code>.</p>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-8_b147a7b17c343e19d62e54f91e546d1a">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset by sample</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>tse_subset_by_sample <span class="ot">&lt;-</span> tse[ , tse<span class="sc">$</span>SampleType <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Feces"</span>, <span class="st">"Skin"</span>, <span class="st">"Tongue"</span>)]</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Show dimensions</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse_subset_by_sample)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 19216     9</code></pre>
</div>
</div>
<p>As a sanity check, the new object <code>tse_subset_by_sample</code> should have the original number of features (rows) and a number of samples (columns) equal to the sum of the samples of interest (in this case 9).</p>
<p>Several characteristics can be used to subset by sample:</p>
<ul>
<li>origin</li>
<li>sampling time</li>
<li>sequencing method</li>
<li>DNA / RNA barcode</li>
<li>cohort</li>
</ul>
</section>
<section id="subset-by-feature-row-wise" class="level4" data-number="5.1.2.2">
<h4 data-number="5.1.2.2" class="anchored" data-anchor-id="subset-by-feature-row-wise"><span class="header-section-number">5.1.2.2</span> Subset by feature (row-wise)</h4>
<p>Similarly, here we will extract a subset containing only the features that belong to the phyla Actinobacteria and Chlamydiae, stored as <code>Phylum</code> within <code>rowData(tse)</code>. However, subsetting by feature implies a few more obstacles, such as the presence of <code>NA</code> elements and the possible need for agglomeration.</p>
<p>As previously, we would first like to see all the possible values that <code>Phylum</code> can take on and how frequent those are:</p>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-9_7b40a925ed254ea1d83498d55d3310de">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Inspect possible values for phylum</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">unique</span>(<span class="fu">rowData</span>(tse)<span class="sc">$</span>Phylum)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] "Crenarchaeota"    "Euryarchaeota"    "Actinobacteria"   "Spirochaetes"    
 [5] "MVP-15"           "Proteobacteria"   "SBR1093"          "Fusobacteria"    
 [9] "Tenericutes"      "ZB3"              "Cyanobacteria"    "GOUTA4"          
[13] "TG3"              "Chlorobi"         "Bacteroidetes"    "Caldithrix"      
[17] "KSB1"             "SAR406"           "LCP-89"           "Thermi"          
[21] "Gemmatimonadetes" "Fibrobacteres"    "GN06"             "AC1"             
[25] "TM6"              "OP8"              "Elusimicrobia"    "NC10"            
[29] "SPAM"             NA                 "Acidobacteria"    "CCM11b"          
[33] "Nitrospirae"      "NKB19"            "BRC1"             "Hyd24-12"        
[37] "WS3"              "PAUC34f"          "GN04"             "GN12"            
[41] "Verrucomicrobia"  "Lentisphaerae"    "LD1"              "Chlamydiae"      
[45] "OP3"              "Planctomycetes"   "Firmicutes"       "OP9"             
[49] "WPS-2"            "Armatimonadetes"  "SC3"              "TM7"             
[53] "GN02"             "SM2F11"           "ABY1_OD1"         "ZB2"             
[57] "OP11"             "Chloroflexi"      "SC4"              "WS1"             
[61] "GAL15"            "AD3"              "WS2"              "Caldiserica"     
[65] "Thermotogae"      "Synergistetes"    "SR1"             </code></pre>
</div>
</div>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-10_7c349a895f0760b70fbbc424bd6cd846">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Show the frequency of each value</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(tse)<span class="sc">$</span>Phylum <span class="sc">%&gt;%</span> <span class="fu">table</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-11_1d06d025dee54172671da1219cc3d357">
<div class="cell-output-display">
<div style="border: 1px solid #ddd; padding: 5px; overflow-x: scroll; width:100%; ">
<table class="table table-striped table-sm small" data-quarto-postprocess="true">
<thead>
<tr class="header">
<th style="text-align: left;" data-quarto-table-cell-role="th">.</th>
<th style="text-align: right;" data-quarto-table-cell-role="th">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ABY1_OD1</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">AC1</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Acidobacteria</td>
<td style="text-align: right;">1021</td>
</tr>
<tr class="even">
<td style="text-align: left;">Actinobacteria</td>
<td style="text-align: right;">1631</td>
</tr>
<tr class="odd">
<td style="text-align: left;">AD3</td>
<td style="text-align: right;">9</td>
</tr>
<tr class="even">
<td style="text-align: left;">Armatimonadetes</td>
<td style="text-align: right;">61</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Bacteroidetes</td>
<td style="text-align: right;">2382</td>
</tr>
<tr class="even">
<td style="text-align: left;">BRC1</td>
<td style="text-align: right;">13</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Caldiserica</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="even">
<td style="text-align: left;">Caldithrix</td>
<td style="text-align: right;">10</td>
</tr>
<tr class="odd">
<td style="text-align: left;">CCM11b</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">Chlamydiae</td>
<td style="text-align: right;">21</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Chlorobi</td>
<td style="text-align: right;">64</td>
</tr>
<tr class="even">
<td style="text-align: left;">Chloroflexi</td>
<td style="text-align: right;">437</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Crenarchaeota</td>
<td style="text-align: right;">106</td>
</tr>
<tr class="even">
<td style="text-align: left;">Cyanobacteria</td>
<td style="text-align: right;">393</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Elusimicrobia</td>
<td style="text-align: right;">31</td>
</tr>
<tr class="even">
<td style="text-align: left;">Euryarchaeota</td>
<td style="text-align: right;">102</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Fibrobacteres</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">Firmicutes</td>
<td style="text-align: right;">4356</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Fusobacteria</td>
<td style="text-align: right;">37</td>
</tr>
<tr class="even">
<td style="text-align: left;">GAL15</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Gemmatimonadetes</td>
<td style="text-align: right;">191</td>
</tr>
<tr class="even">
<td style="text-align: left;">GN02</td>
<td style="text-align: right;">8</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GN04</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">GN06</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">GN12</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">GOUTA4</td>
<td style="text-align: right;">11</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Hyd24-12</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">KSB1</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">LCP-89</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">LD1</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Lentisphaerae</td>
<td style="text-align: right;">21</td>
</tr>
<tr class="even">
<td style="text-align: left;">MVP-15</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NC10</td>
<td style="text-align: right;">9</td>
</tr>
<tr class="even">
<td style="text-align: left;">Nitrospirae</td>
<td style="text-align: right;">74</td>
</tr>
<tr class="odd">
<td style="text-align: left;">NKB19</td>
<td style="text-align: right;">16</td>
</tr>
<tr class="even">
<td style="text-align: left;">OP11</td>
<td style="text-align: right;">6</td>
</tr>
<tr class="odd">
<td style="text-align: left;">OP3</td>
<td style="text-align: right;">30</td>
</tr>
<tr class="even">
<td style="text-align: left;">OP8</td>
<td style="text-align: right;">27</td>
</tr>
<tr class="odd">
<td style="text-align: left;">OP9</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">PAUC34f</td>
<td style="text-align: right;">3</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Planctomycetes</td>
<td style="text-align: right;">638</td>
</tr>
<tr class="even">
<td style="text-align: left;">Proteobacteria</td>
<td style="text-align: right;">6416</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SAR406</td>
<td style="text-align: right;">21</td>
</tr>
<tr class="even">
<td style="text-align: left;">SBR1093</td>
<td style="text-align: right;">9</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SC3</td>
<td style="text-align: right;">8</td>
</tr>
<tr class="even">
<td style="text-align: left;">SC4</td>
<td style="text-align: right;">8</td>
</tr>
<tr class="odd">
<td style="text-align: left;">SM2F11</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">SPAM</td>
<td style="text-align: right;">22</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Spirochaetes</td>
<td style="text-align: right;">124</td>
</tr>
<tr class="even">
<td style="text-align: left;">SR1</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Synergistetes</td>
<td style="text-align: right;">7</td>
</tr>
<tr class="even">
<td style="text-align: left;">Tenericutes</td>
<td style="text-align: right;">143</td>
</tr>
<tr class="odd">
<td style="text-align: left;">TG3</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="even">
<td style="text-align: left;">Thermi</td>
<td style="text-align: right;">46</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Thermotogae</td>
<td style="text-align: right;">1</td>
</tr>
<tr class="even">
<td style="text-align: left;">TM6</td>
<td style="text-align: right;">27</td>
</tr>
<tr class="odd">
<td style="text-align: left;">TM7</td>
<td style="text-align: right;">32</td>
</tr>
<tr class="even">
<td style="text-align: left;">Verrucomicrobia</td>
<td style="text-align: right;">470</td>
</tr>
<tr class="odd">
<td style="text-align: left;">WPS-2</td>
<td style="text-align: right;">20</td>
</tr>
<tr class="even">
<td style="text-align: left;">WS1</td>
<td style="text-align: right;">5</td>
</tr>
<tr class="odd">
<td style="text-align: left;">WS2</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">WS3</td>
<td style="text-align: right;">70</td>
</tr>
<tr class="odd">
<td style="text-align: left;">ZB2</td>
<td style="text-align: right;">2</td>
</tr>
<tr class="even">
<td style="text-align: left;">ZB3</td>
<td style="text-align: right;">2</td>
</tr>
</tbody>
</table>
</div>

</div>
</div>
<p><strong>Note</strong>: after subsetting, expect the number of columns to equal the sum of the frequencies of the feature(s) that you are interested in. For instance, <code>nrows = Actinobacteria + Chlamydiae = 1631 + 21 =   1652</code>.</p>
<p>Depending on your research question, you might or might not need to agglomerate the data in the first place: if you want to find the abundance of each and every feature that belongs to Actinobacteria and Chlamydiae, agglomeration is not needed; if you want to find the total abundance of all features that belong to Actinobacteria or Chlamydiae, agglomeration is recommended.</p>
<section id="non-agglomerated-data" class="level5" data-number="5.1.2.2.1">
<h5 data-number="5.1.2.2.1" class="anchored" data-anchor-id="non-agglomerated-data"><span class="header-section-number">5.1.2.2.1</span> Non-agglomerated data</h5>
<p>Next, we <em>logical index</em> across the rows of <code>tse</code> (make sure to leave the second index empty to select all columns) and filter for the features that fall in either Actinobacteria or Chlamydiae group. For this, we use the information on the samples from the metadata <code>rowData(tse)</code>.</p>
<p>The first term with the <code>%in%</code> operator includes all the features of interest, whereas the second term after the AND operator <code>&amp;</code> filters out all features that have an <code>NA</code> in place of the phylum variable.</p>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-12_5bf5d9b4ec92c5cb747e506b2743b525">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset by feature</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>tse_subset_by_feature <span class="ot">&lt;-</span> tse[<span class="fu">rowData</span>(tse)<span class="sc">$</span>Phylum <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Actinobacteria"</span>, <span class="st">"Chlamydiae"</span>) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(tse)<span class="sc">$</span>Phylum), ]</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Show dimensions</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse_subset_by_feature)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1652   26</code></pre>
</div>
</div>
<p>As a sanity check, the new object, <code>tse_subset_by_feature</code>, should have the original number of samples (columns) and a number of features (rows) equal to the sum of the features of interest (in this case, 1652).</p>
</section>
<section id="agglomerated-data" class="level5" data-number="5.1.2.2.2">
<h5 data-number="5.1.2.2.2" class="anchored" data-anchor-id="agglomerated-data"><span class="header-section-number">5.1.2.2.2</span> Agglomerated data</h5>
<p>When total abundances of certain phyla are of relevance, the data is initially agglomerated by Phylum. Then, similar steps as in the case of non-agglomerated data are followed.</p>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-13_4d64fbe31ab03efec158a36aebbdbee3">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Agglomerate by phylum</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a>tse_phylum <span class="ot">&lt;-</span> tse <span class="sc">%&gt;%</span> <span class="fu">mergeFeaturesByRank</span>(<span class="at">rank =</span> <span class="st">"Phylum"</span>)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset by feature and remove NAs</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>tse_phylum_subset_by_feature <span class="ot">&lt;-</span> tse_phylum[<span class="fu">rowData</span>(tse_phylum)<span class="sc">$</span>Phylum <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Actinobacteria"</span>, <span class="st">"Chlamydiae"</span>) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(tse_phylum)<span class="sc">$</span>Phylum), ]</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Show dimensions</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse_phylum_subset_by_feature)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1]  2 26</code></pre>
</div>
</div>
<p><strong>Note</strong>: as data was agglomerated, the number of rows should equal the number of phyla used to index (in this case, just 2).</p>
<p>Alternatively:</p>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-14_532a0b2e87ee433a03f70f2fa1cbde7f">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Store features of interest into phyla</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>phyla <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Phylum:Actinobacteria"</span>, <span class="st">"Phylum:Chlamydiae"</span>)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a><span class="co"># subset by feature</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>tse_phylum_subset_by_feature <span class="ot">&lt;-</span> tse_phylum[phyla, ]</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Show dimensions</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse_subset_by_feature)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1652   26</code></pre>
</div>
</div>
<p>The code above returns the non-agglomerated version of the data.</p>
<p>Fewer characteristics can be used to subset by feature:</p>
<ul>
<li>Taxonomic rank</li>
<li>Meta-taxonomic group</li>
</ul>
<p>For subsetting by kingdom, agglomeration does not apply, whereas for the other ranks it can be applied if necessary.</p>
</section>
</section>
<section id="subset-by-sample-and-feature" class="level4" data-number="5.1.2.3">
<h4 data-number="5.1.2.3" class="anchored" data-anchor-id="subset-by-sample-and-feature"><span class="header-section-number">5.1.2.3</span> Subset by sample and feature</h4>
<p>Finally, we can subset data by sample and feature at once. The resulting subset contains all the samples of human origin and all the features of phyla Actinobacteria or Chlamydiae.</p>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-15_8f984bbc87f2572d3905b7b0edf4b992">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset by sample and feature and remove NAs</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>tse_subset_by_sample_feature <span class="ot">&lt;-</span> tse[<span class="fu">rowData</span>(tse)<span class="sc">$</span>Phylum <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Actinobacteria"</span>, <span class="st">"Chlamydiae"</span>) <span class="sc">&amp;</span> <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(tse)<span class="sc">$</span>Phylum), tse<span class="sc">$</span>SampleType <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Feces"</span>, <span class="st">"Skin"</span>, <span class="st">"Tongue"</span>)]</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Show dimensions</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(tse_subset_by_sample_feature)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 1652    9</code></pre>
</div>
</div>
<p><strong>Note</strong>: the dimensions of <code>tse_subset_by_sample_feature</code> agree with those of the previous subsets (9 columns filtered by sample and 1652 rows filtered by feature).</p>
<p>If a study was to consider and quantify the presence of Actinobacteria as well as Chlamydiae in different sites of the human body, <code>tse_subset_by_sample_feature</code> might be a suitable subset to start with.</p>
</section>
<section id="remove-empty-columns-and-rows" class="level4" data-number="5.1.2.4">
<h4 data-number="5.1.2.4" class="anchored" data-anchor-id="remove-empty-columns-and-rows"><span class="header-section-number">5.1.2.4</span> Remove empty columns and rows</h4>
<p>Sometimes data might contain, e.g., features that are not present in any of the samples. This can occur, for example, after the data subsetting. In certain analyses, we might want to remove those instances.</p>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-16_e686ac7c2c11c150ed60bc37e2dc974f">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Agglomerate data at Genus level </span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>tse_genus <span class="ot">&lt;-</span> <span class="fu">mergeFeaturesByRank</span>(tse, <span class="at">rank =</span> <span class="st">"Genus"</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a><span class="co"># List bacteria that we want to include</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>genera <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"Class:Thermoprotei"</span>, <span class="st">"Genus:Sulfolobus"</span>, <span class="st">"Genus:Sediminicola"</span>)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset data</span></span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true" tabindex="-1"></a>tse_genus_sub <span class="ot">&lt;-</span> tse_genus[genera, ]</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true" tabindex="-1"></a>tse_genus_sub</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: TreeSummarizedExperiment 
dim: 3 26 
metadata(1): agglomerated_by_rank
assays(1): counts
rownames(3): Class:Thermoprotei Genus:Sulfolobus Genus:Sediminicola
rowData names(7): Kingdom Phylum ... Genus Species
colnames(26): CL3 CC1 ... Even2 Even3
colData names(7): X.SampleID Primer ... SampleType Description
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: a LinkDataFrame (3 rows)
rowTree: 1 phylo tree(s) (19216 leaves)
colLinks: NULL
colTree: NULL</code></pre>
</div>
</div>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-17_07204784528c8fb7d3940339142b57b8">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># List total counts of each sample</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colSums</span>(<span class="fu">assay</span>(tse_genus_sub, <span class="st">"counts"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     CL3      CC1      SV1  M31Fcsw  M11Fcsw  M31Plmr  M11Plmr  F21Plmr 
       1        0        0        1        1        0        4        1 
 M31Tong  M11Tong LMEpi24M SLEpi20M   AQC1cm   AQC4cm   AQC7cm      NP2 
       7        3        0        2       64      105      136      222 
     NP3      NP5  TRRsed1  TRRsed2  TRRsed3     TS28     TS29    Even1 
    6433     1154        2        2        2        0        0        0 
   Even2    Even3 
       2        0 </code></pre>
</div>
</div>
<p>Now we can see that certain samples do not include any bacteria. We can remove those.</p>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-18_b88ee7fa85df372ad598bd0b83c54e06">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Remove samples that do not contain any bacteria</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>tse_genus_sub <span class="ot">&lt;-</span> tse_genus_sub[ , <span class="fu">colSums</span>(<span class="fu">assay</span>(tse_genus_sub, <span class="st">"counts"</span>)) <span class="sc">!=</span> <span class="dv">0</span> ]</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>tse_genus_sub</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: TreeSummarizedExperiment 
dim: 3 18 
metadata(1): agglomerated_by_rank
assays(1): counts
rownames(3): Class:Thermoprotei Genus:Sulfolobus Genus:Sediminicola
rowData names(7): Kingdom Phylum ... Genus Species
colnames(18): CL3 M31Fcsw ... TRRsed3 Even2
colData names(7): X.SampleID Primer ... SampleType Description
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: a LinkDataFrame (3 rows)
rowTree: 1 phylo tree(s) (19216 leaves)
colLinks: NULL
colTree: NULL</code></pre>
</div>
</div>
<p>The same action can also be applied to the features.</p>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-19_76e4e1f331f9b4f2d555c759e5f0d7d2">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take only those samples that are collected from feces, skin, or tongue</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>tse_genus_sub <span class="ot">&lt;-</span> tse_genus[ , tse_genus<span class="sc">$</span>SampleType <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"Feces"</span>, <span class="st">"Skin"</span>, <span class="st">"Tongue"</span>)]</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>tse_genus_sub</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: TreeSummarizedExperiment 
dim: 1516 9 
metadata(1): agglomerated_by_rank
assays(1): counts
rownames(1516): Class:Thermoprotei Genus:Sulfolobus ...
  Genus:Coprothermobacter Phylum:SR1
rowData names(7): Kingdom Phylum ... Genus Species
colnames(9): M31Fcsw M11Fcsw ... TS28 TS29
colData names(7): X.SampleID Primer ... SampleType Description
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: a LinkDataFrame (1516 rows)
rowTree: 1 phylo tree(s) (19216 leaves)
colLinks: NULL
colTree: NULL</code></pre>
</div>
</div>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-20_e72f98e45c34344b6a81111633d367d6">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a><span class="co"># What is the number of bacteria that are not present?</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a><span class="fu">sum</span>(<span class="fu">rowSums</span>(<span class="fu">assay</span>(tse_genus_sub, <span class="st">"counts"</span>)) <span class="sc">==</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 435</code></pre>
</div>
</div>
<p>We can see that there are bacteria that are not present in these samples we chose. We can remove those bacteria from the data.</p>
<div class="cell" data-hash="10_manipulation_cache/html/unnamed-chunk-21_fa06ddb5b1c2c276e1e003354be72060">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take only those bacteria that are present</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>tse_genus_sub <span class="ot">&lt;-</span> tse_genus_sub[<span class="fu">rowSums</span>(<span class="fu">assay</span>(tse_genus_sub, <span class="st">"counts"</span>)) <span class="sc">&gt;</span> <span class="dv">0</span>, ]</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a>tse_genus_sub</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: TreeSummarizedExperiment 
dim: 1081 9 
metadata(1): agglomerated_by_rank
assays(1): counts
rownames(1081): Genus:Sulfolobus Order:NRP-J ...
  Genus:Coprothermobacter Phylum:SR1
rowData names(7): Kingdom Phylum ... Genus Species
colnames(9): M31Fcsw M11Fcsw ... TS28 TS29
colData names(7): X.SampleID Primer ... SampleType Description
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: a LinkDataFrame (1081 rows)
rowTree: 1 phylo tree(s) (19216 leaves)
colLinks: NULL
colTree: NULL</code></pre>
</div>
</div>
</section>
</section>
<section id="splitting" class="level3" data-number="5.1.3">
<h3 data-number="5.1.3" class="anchored" data-anchor-id="splitting"><span class="header-section-number">5.1.3</span> Splitting</h3>
<p>You can split the data based on variables by using the functions <code>splitByRanks</code> and <code>splitOn</code>.</p>
<p><code>splitByRanks</code> splits the data based on taxonomic ranks. Since the elements of the output list share columns, they can be stored into <code>altExp</code>.</p>
<div class="cell" data-hash="10_manipulation_cache/html/splitbyRanks_043e984ebda19d69b5d2a89d4a9be080">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">altExps</span>(tse) <span class="ot">&lt;-</span> <span class="fu">splitByRanks</span>(tse)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">altExps</span>(tse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 7
names(7): Kingdom Phylum Class Order Family Genus Species</code></pre>
</div>
</div>
<p>If you want to split the data based on another variable than taxonomic rank, use <code>splitOn</code>. It works for row-wise and column-wise splitting.</p>
<div class="cell" data-hash="10_manipulation_cache/html/splitOn_526de5606d63bee6c9d1de25dcc94739">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">splitOn</span>(tse, <span class="st">"SampleType"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 9
names(9): Soil Feces Skin Tongue ... Ocean Sediment (estuary) Mock</code></pre>
</div>
</div>
</section>
</section>
<section id="add-or-modify-data" class="level2" data-number="5.2">
<h2 data-number="5.2" class="anchored" data-anchor-id="add-or-modify-data"><span class="header-section-number">5.2</span> Add or modify data</h2>
<p>The information contained by the <code>colData</code> of a <code>TreeSE</code> can be modified by accessing the desired variables.</p>
<div class="cell" data-hash="10_manipulation_cache/html/modify-coldata_44789a7c015f22da69ea411fec9d38f9">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="co"># modify the Description entries</span></span>
<span id="cb37-2"><a href="#cb37-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(tse)<span class="sc">$</span>Description <span class="ot">&lt;-</span> <span class="fu">paste</span>(<span class="fu">colData</span>(tse)<span class="sc">$</span>Description, <span class="st">"modified description"</span>)</span>
<span id="cb37-3"><a href="#cb37-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb37-4"><a href="#cb37-4" aria-hidden="true" tabindex="-1"></a><span class="co"># view modified variable</span></span>
<span id="cb37-5"><a href="#cb37-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tse<span class="sc">$</span>Description)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Calhoun South Carolina Pine soil, pH 4.9 modified description"  
[2] "Cedar Creek Minnesota, grassland, pH 6.1 modified description"  
[3] "Sevilleta new Mexico, desert scrub, pH 8.3 modified description"
[4] "M3, Day 1, fecal swab, whole body study modified description"   
[5] "M1, Day 1, fecal swab, whole body study  modified description"  
[6] "M3, Day 1, right palm, whole body study modified description"   </code></pre>
</div>
</div>
<p>New information can also be added to the experiment by creating a new variable.</p>
<div class="cell" data-hash="10_manipulation_cache/html/add-coldata_4f876787acf5acfe91b2dbb6ec333d9a">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># simulate new data</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>new_data <span class="ot">&lt;-</span> <span class="fu">runif</span>(<span class="fu">ncol</span>(tse))</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a><span class="co"># store new data as new variable in colData</span></span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(tse)<span class="sc">$</span>NewVariable <span class="ot">&lt;-</span> new_data</span>
<span id="cb39-6"><a href="#cb39-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb39-7"><a href="#cb39-7" aria-hidden="true" tabindex="-1"></a><span class="co"># view new variable</span></span>
<span id="cb39-8"><a href="#cb39-8" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(tse<span class="sc">$</span>NewVariable)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0.8597 0.8657 0.3489 0.1964 0.9575 0.1664</code></pre>
</div>
</div>
</section>
<section id="merge-data" class="level2" data-number="5.3">
<h2 data-number="5.3" class="anchored" data-anchor-id="merge-data"><span class="header-section-number">5.3</span> Merge data</h2>
<p><code>mia</code> package has <code>mergeSEs</code> function that merges multiple <code>SummarizedExperiment</code> objects. For example, it is possible to combine multiple <code>TreeSE</code> objects which each includes one sample.</p>
<p><code>mergeSEs</code> works like <code>dplyr</code> joining functions. In fact, there are available <code>dplyr-like</code> aliases of <code>mergeSEs</code>, such as <code>full_join</code>.</p>
<div class="cell" data-hash="10_manipulation_cache/html/merge1_63c06c7f035d6da619ae220ffe9ec63b">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Take subsets for demonstration purposes</span></span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>tse1 <span class="ot">&lt;-</span> tse[, <span class="dv">1</span>]</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>tse2 <span class="ot">&lt;-</span> tse[, <span class="dv">2</span>]</span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>tse3 <span class="ot">&lt;-</span> tse[, <span class="dv">3</span>]</span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>tse4 <span class="ot">&lt;-</span> tse[<span class="dv">1</span><span class="sc">:</span><span class="dv">100</span>, <span class="dv">4</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-hash="10_manipulation_cache/html/merge2_e0bcfe6efbae8e8b76faa042e127d86e">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># With inner join, we want to include all shared rows. When using mergeSEs function</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a><span class="co"># all samples are always preserved.</span></span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">mergeSEs</span>(<span class="fu">list</span>(tse1, tse2, tse3, tse4), <span class="at">join =</span> <span class="st">"inner"</span>)</span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>tse</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: TreeSummarizedExperiment 
dim: 100 4 
metadata(0):
assays(1): counts
rownames(100): 239672 243675 ... 549322 951
rowData names(7): Kingdom Phylum ... Genus Species
colnames(4): CC1 CL3 M31Fcsw SV1
colData names(8): X.SampleID Primer ... Description NewVariable
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: a LinkDataFrame (100 rows)
rowTree: 1 phylo tree(s) (19216 leaves)
colLinks: NULL
colTree: NULL</code></pre>
</div>
</div>
<div class="cell" data-hash="10_manipulation_cache/html/merge3_3be22a1da326eac33a526429d969ad01">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Left join preserves all rows of the 1st object</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> mia<span class="sc">::</span><span class="fu">left_join</span>(tse1, tse4, <span class="at">missing_values =</span> <span class="dv">0</span>)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>tse</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: TreeSummarizedExperiment 
dim: 19216 2 
metadata(0):
assays(1): counts
rownames(19216): 239672 243675 ... 239967 254851
rowData names(7): Kingdom Phylum ... Genus Species
colnames(2): CL3 M31Fcsw
colData names(8): X.SampleID Primer ... Description NewVariable
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: a LinkDataFrame (19216 rows)
rowTree: 1 phylo tree(s) (19216 leaves)
colLinks: NULL
colTree: NULL</code></pre>
</div>
</div>
<section id="additional-functions" class="level3" data-number="5.3.1">
<h3 data-number="5.3.1" class="anchored" data-anchor-id="additional-functions"><span class="header-section-number">5.3.1</span> Additional functions</h3>
<ul>
<li><a href="https://microbiome.github.io/mia/reference/taxonomy-methods.html">mapTaxonomy</a></li>
<li><a href="https://microbiome.github.io/mia/reference/merge-methods.html">mergeFeatures/mergeSamples</a></li>
</ul>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./04_containers.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./12_quality_control.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>