<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>7&nbsp; Taxonomic Information</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
/* CSS for citations */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging-indent div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./14_alpha_diversity.html" rel="next">
<link href="./12_quality_control.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./11_taxonomic_information.html"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Orchestrating Microbiome Analysis with Bioconductor</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="" rel="" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-twitter"></i></a>
    <a href="https://microbiome.github.io/mia/" rel="" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-facebook"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Orchestrating Microbiome Analysis with Bioconductor</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./11_taxonomic_information.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Similarity</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./24_clustering.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Community Typing (Clustering)</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./30_differential_abundance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">18</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">19</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Developers</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Sessioninfo</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#assigning-taxonomic-information." id="toc-assigning-taxonomic-information." class="nav-link active" data-scroll-target="#assigning-taxonomic-information."><span class="header-section-number">7.1</span> Assigning taxonomic information.</a>
  <ul class="collapse">
  <li><a href="#dada2" id="toc-dada2" class="nav-link" data-scroll-target="#dada2"><span class="header-section-number">7.1.1</span> dada2</a></li>
  <li><a href="#decipher" id="toc-decipher" class="nav-link" data-scroll-target="#decipher"><span class="header-section-number">7.1.2</span> DECIPHER</a></li>
  </ul></li>
  <li><a href="#functions-to-access-taxonomic-information" id="toc-functions-to-access-taxonomic-information" class="nav-link" data-scroll-target="#functions-to-access-taxonomic-information"><span class="header-section-number">7.2</span> Functions to access taxonomic information</a>
  <ul class="collapse">
  <li><a href="#fly-tree" id="toc-fly-tree" class="nav-link" data-scroll-target="#fly-tree"><span class="header-section-number">7.2.1</span> Generate a taxonomic tree on the fly</a></li>
  </ul></li>
  <li><a href="#data-agglomeration" id="toc-data-agglomeration" class="nav-link" data-scroll-target="#data-agglomeration"><span class="header-section-number">7.3</span> Data agglomeration</a>
  <ul class="collapse">
  <li><a href="#taxa-clustering" id="toc-taxa-clustering" class="nav-link" data-scroll-target="#taxa-clustering"><span class="header-section-number">7.3.1</span> Taxa clustering</a></li>
  </ul></li>
  <li><a href="#assay-transform" id="toc-assay-transform" class="nav-link" data-scroll-target="#assay-transform"><span class="header-section-number">7.4</span> Data transformation</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="taxonomic-information" class="quarto-section-identifier"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Taxonomic Information</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-2_88297826ce97fbd4157bfe88e0821c6f">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"GlobalPatterns"</span>, <span class="at">package =</span> <span class="st">"mia"</span>)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> GlobalPatterns</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Taxonomic information is a key part of analyzing microbiome data and without it, any type of data analysis probably will not make much sense. However, the degree of detail of taxonomic information differs depending on the dataset and annotation data used.</p>
<p>Therefore, the mia package expects a loose assembly of taxonomic information and assumes certain key aspects:</p>
<ul>
<li>Taxonomic information is given as character vectors or factors in the <code>rowData</code> of a <code>SummarizedExperiment</code> object.</li>
<li>The columns containing the taxonomic information must be named <code>domain</code>, <code>kingdom</code>, <code>phylum</code>, <code>class</code>, <code>order</code>, <code>family</code>, <code>genus</code>, <code>species</code> or with a capital first letter.</li>
<li>the columns must be given in the order shown above</li>
<li>column can be omited, but the order must remain</li>
</ul>
<p>In this chapter, we will refer to co-abundant groups as CAGs, which are clusters of taxa that co-vary across samples.</p>
<section id="assigning-taxonomic-information." class="level2" data-number="7.1">
<h2 data-number="7.1" class="anchored" data-anchor-id="assigning-taxonomic-information."><span class="header-section-number">7.1</span> Assigning taxonomic information.</h2>
<p>There are a number of methods to assign taxonomic information. We like to give a short introduction about the methods available without ranking one over the other. This has to be your choice based on the result for the individual dataset.</p>
<section id="dada2" class="level3" data-number="7.1.1">
<h3 data-number="7.1.1" class="anchored" data-anchor-id="dada2"><span class="header-section-number">7.1.1</span> dada2</h3>
<p>The dada2 package <span class="citation" data-cites="Callahan2016dada2">(<a href="#ref-Callahan2016dada2" role="doc-biblioref">Callahan et al. 2016</a>)</span> implements the <code>assignTaxonomy</code> function, which takes as input the ASV sequences associated with each row of data and a training dataset. For more information visit the <a href="https://benjjneb.github.io/dada2/assign.html">dada2 homepage</a>.</p>
</section>
<section id="decipher" class="level3" data-number="7.1.2">
<h3 data-number="7.1.2" class="anchored" data-anchor-id="decipher"><span class="header-section-number">7.1.2</span> DECIPHER</h3>
<p>The DECIPHER package <span class="citation" data-cites="R_DECIPHER">(<a href="#ref-R_DECIPHER" role="doc-biblioref">Wright 2020</a>)</span> implements the <code>IDTAXA</code> algorithm to assign either taxonomic information or function information. For <code>mia</code> only the first option is of interest for now and more information can be found on the <a href="http://www2.decipher.codes/Classification.html">DECIPHER website</a>.</p>
</section>
</section>
<section id="functions-to-access-taxonomic-information" class="level2" data-number="7.2">
<h2 data-number="7.2" class="anchored" data-anchor-id="functions-to-access-taxonomic-information"><span class="header-section-number">7.2</span> Functions to access taxonomic information</h2>
<p><code>checkTaxonomy</code> checks whether the taxonomic information is usable for <code>mia</code></p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-3_2309f335dbca97566e4bd5d4046bf77d">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="fu">checkTaxonomy</span>(tse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
</div>
<p>Since the <code>rowData</code> can contain other data, <code>taxonomyRanks</code> will return the columns <code>mia</code> assumes to contain the taxonomic information.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-4_d92bf3f326f9d2eef3ad0c4871cb6d0f">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">taxonomyRanks</span>(tse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Kingdom" "Phylum"  "Class"   "Order"   "Family"  "Genus"   "Species"</code></pre>
</div>
</div>
<p>This can then be used to subset the <code>rowData</code> to columns needed.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-5_4014c53a6cc20763186a73ee8fa4ab7f">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(tse)[, <span class="fu">taxonomyRanks</span>(tse)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 19216 rows and 7 columns
           Kingdom        Phylum        Class        Order        Family
       &lt;character&gt;   &lt;character&gt;  &lt;character&gt;  &lt;character&gt;   &lt;character&gt;
549322     Archaea Crenarchaeota Thermoprotei           NA            NA
522457     Archaea Crenarchaeota Thermoprotei           NA            NA
951        Archaea Crenarchaeota Thermoprotei Sulfolobales Sulfolobaceae
244423     Archaea Crenarchaeota        Sd-NA           NA            NA
586076     Archaea Crenarchaeota        Sd-NA           NA            NA
...            ...           ...          ...          ...           ...
278222    Bacteria           SR1           NA           NA            NA
463590    Bacteria           SR1           NA           NA            NA
535321    Bacteria           SR1           NA           NA            NA
200359    Bacteria           SR1           NA           NA            NA
271582    Bacteria           SR1           NA           NA            NA
             Genus                Species
       &lt;character&gt;            &lt;character&gt;
549322          NA                     NA
522457          NA                     NA
951     Sulfolobus Sulfolobusacidocalda..
244423          NA                     NA
586076          NA                     NA
...            ...                    ...
278222          NA                     NA
463590          NA                     NA
535321          NA                     NA
200359          NA                     NA
271582          NA                     NA</code></pre>
</div>
</div>
<p><code>taxonomyRankEmpty</code> checks for empty values in the given <code>rank</code> and returns a logical vector of <code>length(x)</code>.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-6_eacb578f573b6e36cea5c13255b4b8ef">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">all</span>(<span class="sc">!</span><span class="fu">taxonomyRankEmpty</span>(tse, <span class="at">rank =</span> <span class="st">"Kingdom"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] TRUE</code></pre>
</div>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">taxonomyRankEmpty</span>(tse, <span class="at">rank =</span> <span class="st">"Genus"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
FALSE  TRUE 
 8008 11208 </code></pre>
</div>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(<span class="fu">taxonomyRankEmpty</span>(tse, <span class="at">rank =</span> <span class="st">"Species"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
FALSE  TRUE 
 1413 17803 </code></pre>
</div>
</div>
<p><code>getTaxonomyLabels</code> is a multi-purpose function, which turns taxonomic information into a character vector of <code>length(x)</code></p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-7_a4646b591faf1ce3b5600473b2d15a34">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getTaxonomyLabels</span>(tse))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Class:Thermoprotei"               "Class:Thermoprotei_1"            
[3] "Species:Sulfolobusacidocaldarius" "Class:Sd-NA"                     
[5] "Class:Sd-NA_1"                    "Class:Sd-NA_2"                   </code></pre>
</div>
</div>
<p>By default, this will use the lowest non-empty information to construct a string with the following scheme <code>level:value</code>. If all levels are the same, this part is omitted, but can be added by setting <code>with_rank = TRUE</code>.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-8_f63678b5fec0fb231fcfdaacfc3697a3">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>phylum <span class="ot">&lt;-</span> <span class="sc">!</span><span class="fu">is.na</span>(<span class="fu">rowData</span>(tse)<span class="sc">$</span>Phylum) <span class="sc">&amp;</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">vapply</span>(<span class="fu">data.frame</span>(<span class="fu">apply</span>(<span class="fu">rowData</span>(tse)[, <span class="fu">taxonomyRanks</span>(tse)[<span class="dv">3</span><span class="sc">:</span><span class="dv">7</span>]], 1L, is.na)), all, <span class="fu">logical</span>(<span class="dv">1</span>))</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getTaxonomyLabels</span>(tse[phylum,]))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Crenarchaeota"    "Crenarchaeota_1"  "Crenarchaeota_2"  "Actinobacteria"  
[5] "Actinobacteria_1" "Spirochaetes"    </code></pre>
</div>
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getTaxonomyLabels</span>(tse[phylum,], <span class="at">with_rank =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Phylum:Crenarchaeota"    "Phylum:Crenarchaeota_1" 
[3] "Phylum:Crenarchaeota_2"  "Phylum:Actinobacteria"  
[5] "Phylum:Actinobacteria_1" "Phylum:Spirochaetes"    </code></pre>
</div>
</div>
<p>By default the return value of <code>getTaxonomyLabels</code> contains only unique elements by passing it through <code>make.unique</code>. This step can be omitted by setting <code>make_unique = FALSE</code>.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-9_4817bd27e10f390735a8e6d6d6d9f848">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getTaxonomyLabels</span>(tse[phylum,], <span class="at">with_rank =</span> <span class="cn">TRUE</span>, <span class="at">make_unique =</span> <span class="cn">FALSE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Phylum:Crenarchaeota"  "Phylum:Crenarchaeota"  "Phylum:Crenarchaeota" 
[4] "Phylum:Actinobacteria" "Phylum:Actinobacteria" "Phylum:Spirochaetes"  </code></pre>
</div>
</div>
<p>To apply the loop resolving function <code>resolveLoop</code> from the <code>TreeSummarizedExperiment</code> package <span class="citation" data-cites="R_TreeSummarizedExperiment">(<a href="#ref-R_TreeSummarizedExperiment" role="doc-biblioref">Huang 2020</a>)</span> within <code>getTaxonomyLabels</code>, set <code>resolve_loops = TRUE</code>.</p>
<p>The function <code>getUniqueFeatures</code> gives a list of unique taxa for the specified taxonomic rank.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-10_8e710fc854ba90ce5466b1064db242e3">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getUniqueFeatures</span>(tse, <span class="at">rank =</span> <span class="st">"Phylum"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Crenarchaeota"  "Euryarchaeota"  "Actinobacteria" "Spirochaetes"  
[5] "MVP-15"         "Proteobacteria"</code></pre>
</div>
</div>
<section id="fly-tree" class="level3" data-number="7.2.1">
<h3 data-number="7.2.1" class="anchored" data-anchor-id="fly-tree"><span class="header-section-number">7.2.1</span> Generate a taxonomic tree on the fly</h3>
<p>To create a taxonomic tree, <code>taxonomyTree</code> used the information and returns a <code>phylo</code> object. Duplicate information from the <code>rowData</code> is removed.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-11_e612b301c42c087a2e7d37fa4f9a13f8">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">taxonomyTree</span>(tse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Phylogenetic tree with 1645 tips and 1089 internal nodes.

Tip labels:
  Species:Cenarchaeumsymbiosum, Species:pIVWA5, Species:CandidatusNitrososphaeragargensis, Species:SCA1145, Species:SCA1170, Species:Sulfolobusacidocaldarius, ...
Node labels:
  root:ALL, Kingdom:Archaea, Phylum:Crenarchaeota, Class:C2, Class:Sd-NA, Class:Thaumarchaeota, ...

Rooted; includes branch lengths.</code></pre>
</div>
</div>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-12_9d6cd8ddb00e1611dcd6c20a99aec334">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">addTaxonomyTree</span>(tse)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>tse</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: TreeSummarizedExperiment 
dim: 19216 26 
metadata(0):
assays(1): counts
rownames(19216): Class:Thermoprotei Class:Thermoprotei ... Phylum:SR1
  Phylum:SR1
rowData names(7): Kingdom Phylum ... Genus Species
colnames(26): CL3 CC1 ... Even2 Even3
colData names(7): X.SampleID Primer ... SampleType Description
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: a LinkDataFrame (19216 rows)
rowTree: 1 phylo tree(s) (1645 leaves)
colLinks: NULL
colTree: NULL</code></pre>
</div>
</div>
<p>The implementation is based on the <code>toTree</code> function from the <code>TreeSummarizedExperiment</code> package <span class="citation" data-cites="R_TreeSummarizedExperiment">(<a href="#ref-R_TreeSummarizedExperiment" role="doc-biblioref">Huang 2020</a>)</span>.</p>
</section>
</section>
<section id="data-agglomeration" class="level2" data-number="7.3">
<h2 data-number="7.3" class="anchored" data-anchor-id="data-agglomeration"><span class="header-section-number">7.3</span> Data agglomeration</h2>
<p>One of the main applications of taxonomic information in regards to count data is to agglomerate count data on taxonomic levels and track the influence of changing conditions through these levels. For this <code>mia</code> contains the <code>mergeFeaturesByRank</code> function. The ideal location to store the agglomerated data is as an alternative experiment.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-13_0c11ad14b12035a59cd3b965d5830117">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">assay.type =</span> <span class="st">"counts"</span>, <span class="at">method =</span> <span class="st">"relabundance"</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="fu">altExp</span>(tse, <span class="st">"Family"</span>) <span class="ot">&lt;-</span> <span class="fu">mergeFeaturesByRank</span>(tse, <span class="at">rank =</span> <span class="st">"Family"</span>,</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>                                           <span class="at">agglomerateTree =</span> <span class="cn">TRUE</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a><span class="fu">altExp</span>(tse, <span class="st">"Family"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: TreeSummarizedExperiment 
dim: 603 26 
metadata(1): agglomerated_by_rank
assays(2): counts relabundance
rownames(603): Class:Thermoprotei Family:Sulfolobaceae ...
  Family:Thermodesulfobiaceae Phylum:SR1
rowData names(7): Kingdom Phylum ... Genus Species
colnames(26): CL3 CC1 ... Even2 Even3
colData names(7): X.SampleID Primer ... SampleType Description
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: a LinkDataFrame (603 rows)
rowTree: 1 phylo tree(s) (496 leaves)
colLinks: NULL
colTree: NULL</code></pre>
</div>
</div>
<p>If multiple assays (counts and relabundance) exist, both will be agglomerated.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-14_e266864ba1db19a6b37c36dbbe087955">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assayNames</span>(tse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "counts"       "relabundance"</code></pre>
</div>
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assayNames</span>(<span class="fu">altExp</span>(tse, <span class="st">"Family"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "counts"       "relabundance"</code></pre>
</div>
</div>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-15_be5c0a892ff3d1a6bb41c374005c53ee">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(<span class="fu">altExp</span>(tse, <span class="st">"Family"</span>), <span class="st">"relabundance"</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                           CL3       CC1 SV1 M31Fcsw M11Fcsw M31Plmr   M11Plmr
Class:Thermoprotei   0.0000000 0.000e+00   0       0       0       0 0.000e+00
Family:Sulfolobaceae 0.0000000 0.000e+00   0       0       0       0 2.305e-06
Class:Sd-NA          0.0000000 0.000e+00   0       0       0       0 0.000e+00
Order:NRP-J          0.0001991 2.070e-04   0       0       0       0 6.914e-06
Family:SAGMA-X       0.0000000 6.165e-06   0       0       0       0 0.000e+00</code></pre>
</div>
</div>
<div class="cell" data-hash="11_taxonomic_information_cache/html/taxinfo_altexp_example_8935397d5431d99df3e13986f3a46581">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(<span class="fu">altExp</span>(tse, <span class="st">"Family"</span>), <span class="st">"counts"</span>)[<span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                     CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr
Class:Thermoprotei     0   0   0       0       0       0       0
Family:Sulfolobaceae   0   0   0       0       0       0       1
Class:Sd-NA            0   0   0       0       0       0       0
Order:NRP-J          172 235   0       0       0       0       3
Family:SAGMA-X         0   7   0       0       0       0       0</code></pre>
</div>
</div>
<p><code>altExpNames</code> now consists of <code>Family</code> level data. This can be extended to use any taxonomic level listed in <code>mia::taxonomyRanks(tse)</code>.</p>
<p>Rare taxa can also be aggregated into a single group “Other” instead of filtering them out. A suitable function for this is <code>mergeFeaturesByPrevalence</code>. The number of rare taxa is higher on the species level, which causes the need for data agglomeration by prevalence.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-17_ba014f38a568dd01daff9fc4ccd0189d">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">altExp</span>(tse, <span class="st">"Species_byPrevalence"</span>) <span class="ot">&lt;-</span> <span class="fu">mergeFeaturesByPrevalence</span>(tse, </span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>                                                               <span class="at">rank =</span> <span class="st">"Species"</span>, </span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>                                                               <span class="at">other_label =</span> <span class="st">"Other"</span>, </span>
<span id="cb38-4"><a href="#cb38-4" aria-hidden="true" tabindex="-1"></a>                                                               <span class="at">prevalence =</span> <span class="dv">5</span> <span class="sc">/</span> <span class="dv">100</span>, </span>
<span id="cb38-5"><a href="#cb38-5" aria-hidden="true" tabindex="-1"></a>                                                               <span class="at">detection =</span> <span class="dv">1</span> <span class="sc">/</span> <span class="dv">100</span>, </span>
<span id="cb38-6"><a href="#cb38-6" aria-hidden="true" tabindex="-1"></a>                                                               <span class="at">as_relative =</span> T)</span>
<span id="cb38-7"><a href="#cb38-7" aria-hidden="true" tabindex="-1"></a><span class="fu">altExp</span>(tse, <span class="st">"Species_byPrevalence"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: TreeSummarizedExperiment 
dim: 92 26 
metadata(2): agglomerated_by_rank agglomerated_by_rank
assays(2): counts relabundance
rownames(92): pIVWA5 SCA1145 ... Desulfitobacteriumhafniense Other
rowData names(7): Kingdom Phylum ... Genus Species
colnames(26): CL3 CC1 ... Even2 Even3
colData names(7): X.SampleID Primer ... SampleType Description
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: NULL
rowTree: NULL
colLinks: NULL
colTree: NULL</code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">assay</span>(<span class="fu">altExp</span>(tse, <span class="st">"Species_byPrevalence"</span>), <span class="st">"relabundance"</span>)[<span class="dv">88</span><span class="sc">:</span><span class="dv">92</span>, <span class="dv">1</span><span class="sc">:</span><span class="dv">7</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                  CL3       CC1       SV1   M31Fcsw   M11Fcsw
Streptococcusthermophilus   5.787e-06 2.290e-05 1.290e-05 6.032e-04 1.122e-04
Mitsuokellamultacida        8.101e-06 7.046e-06 1.147e-05 6.479e-07 9.632e-07
Veillonellaparvula          1.736e-05 1.673e-05 1.720e-05 7.645e-05 1.589e-05
Desulfitobacteriumhafniense 1.620e-05 1.585e-05 8.602e-06 1.296e-06 9.632e-07
Other                       8.622e-03 6.787e-03 4.325e-02 2.763e-02 2.682e-03
                              M31Plmr  M11Plmr
Streptococcusthermophilus   1.225e-02 0.002478
Mitsuokellamultacida        2.782e-06 0.000000
Veillonellaparvula          2.075e-02 0.001143
Desulfitobacteriumhafniense 0.000e+00 0.000000
Other                       7.077e-02 0.070752</code></pre>
</div>
</div>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-18_031ed3e2d252f25278387a357b45721b">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Saving the tse for later</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>tseGlobalPatterns <span class="ot">&lt;-</span> tse</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="taxa-clustering" class="level3" data-number="7.3.1">
<h3 data-number="7.3.1" class="anchored" data-anchor-id="taxa-clustering"><span class="header-section-number">7.3.1</span> Taxa clustering</h3>
<p>Another way to agglomerate the data is to cluster the taxa. To do so, we usually start by doing a compositionality aware transformation such as CLR, followed by the application of a standard clustering method.</p>
<p>Here is an example that does a CLR transformation followed by the hierarchical clustering algorithm.</p>
<p>First, we import the library <code>bluster</code> that simplifies the clustering.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/bluster_dependence_e2132701919cffbc9d1a73b9e678cbce">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(bluster)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Then we do the CLR transform followed by the clustering. We will cluster with two different distances: the euclidean distance and the kendall distance.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/taxa_clustering_33ec366ea3043931379018dffec4b13f">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the data</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"peerj13075"</span>, <span class="at">package =</span> <span class="st">"mia"</span>)</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> peerj13075</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a><span class="co"># The result of the CLR transform is stored in the assay clr</span></span>
<span id="cb44-6"><a href="#cb44-6" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">method =</span> <span class="st">"clr"</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb44-7"><a href="#cb44-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-8"><a href="#cb44-8" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">assay.type =</span> <span class="st">"clr"</span>, <span class="at">method =</span> <span class="st">"z"</span>, </span>
<span id="cb44-9"><a href="#cb44-9" aria-hidden="true" tabindex="-1"></a>                      <span class="at">MARGIN =</span> <span class="st">"features"</span>)</span>
<span id="cb44-10"><a href="#cb44-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-11"><a href="#cb44-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Cluster (with euclidean distance) on the features of the z assay</span></span>
<span id="cb44-12"><a href="#cb44-12" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">cluster</span>(tse,</span>
<span id="cb44-13"><a href="#cb44-13" aria-hidden="true" tabindex="-1"></a>               <span class="at">assay.type =</span> <span class="st">"z"</span>,</span>
<span id="cb44-14"><a href="#cb44-14" aria-hidden="true" tabindex="-1"></a>               <span class="at">clust.col =</span> <span class="st">"hclustEuclidean"</span>,</span>
<span id="cb44-15"><a href="#cb44-15" aria-hidden="true" tabindex="-1"></a>           <span class="at">MARGIN =</span> <span class="st">"features"</span>,</span>
<span id="cb44-16"><a href="#cb44-16" aria-hidden="true" tabindex="-1"></a>               <span class="fu">HclustParam</span>(<span class="at">dist.fun =</span> stats<span class="sc">::</span>dist, <span class="at">method =</span> <span class="st">"ward.D2"</span>))</span>
<span id="cb44-17"><a href="#cb44-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-18"><a href="#cb44-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Declare the Kendall dissimilarity computation function</span></span>
<span id="cb44-19"><a href="#cb44-19" aria-hidden="true" tabindex="-1"></a>kendall_dissimilarity <span class="ot">&lt;-</span> <span class="cf">function</span>(x) {</span>
<span id="cb44-20"><a href="#cb44-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.dist</span>(<span class="dv">1</span> <span class="sc">-</span> <span class="fu">cor</span>(<span class="fu">t</span>(x), <span class="at">method =</span> <span class="st">"kendall"</span>))</span>
<span id="cb44-21"><a href="#cb44-21" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb44-22"><a href="#cb44-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb44-23"><a href="#cb44-23" aria-hidden="true" tabindex="-1"></a><span class="co"># Cluster (with Kendall dissimilarity) on the features of the z assay</span></span>
<span id="cb44-24"><a href="#cb44-24" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">cluster</span>(tse,</span>
<span id="cb44-25"><a href="#cb44-25" aria-hidden="true" tabindex="-1"></a>               <span class="at">assay.type =</span> <span class="st">"z"</span>,</span>
<span id="cb44-26"><a href="#cb44-26" aria-hidden="true" tabindex="-1"></a>               <span class="at">clust.col =</span> <span class="st">"hclustKendall"</span>,</span>
<span id="cb44-27"><a href="#cb44-27" aria-hidden="true" tabindex="-1"></a>               <span class="at">MARGIN =</span> <span class="st">"features"</span>,            </span>
<span id="cb44-28"><a href="#cb44-28" aria-hidden="true" tabindex="-1"></a>               <span class="fu">HclustParam</span>(<span class="at">dist.fun =</span> kendall_dissimilarity, <span class="at">method =</span> <span class="st">"ward.D2"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Let us store the resulting cluster indices in the <code>rowData</code> column specified with the <code>clust.col</code> parameter.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/taxa_clustering_result_30ba1e78b7578d304375d30562a170d7">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Checking the clusters</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>clusters_euclidean <span class="ot">&lt;-</span> <span class="fu">rowData</span>(tse)<span class="sc">$</span>hclustEuclidean</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(clusters_euclidean, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> OTU1  OTU2  OTU7  OTU9 OTU10 OTU12 OTU14 OTU15 OTU18 OTU19 
    1     2     1     1     1     1     3     4     3     2 
Levels: 1 2 3 4 5</code></pre>
</div>
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>clusters_kendall <span class="ot">&lt;-</span> <span class="fu">rowData</span>(tse)<span class="sc">$</span>hclustKendall</span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(clusters_kendall, <span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> OTU1  OTU2  OTU7  OTU9 OTU10 OTU12 OTU14 OTU15 OTU18 OTU19 
    1     2     1     3     3     1     3     1     1     3 
Levels: 1 2 3 4</code></pre>
</div>
</div>
<p>To better visualize the results and the distribution of the clusters, we can plot the histogram of the clusters.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/taxa_clustering_histogram_4d0ecb5278afc850e3f78bb2b245d03e">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork) <span class="co"># TO arrange several plots as a grid</span></span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>plot1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(<span class="fu">rowData</span>(tse)), <span class="fu">aes</span>(<span class="at">x =</span> clusters_euclidean)) <span class="sc">+</span></span>
<span id="cb49-4"><a href="#cb49-4" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb49-5"><a href="#cb49-5" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"CAG size distribution (Euclidean distance)"</span>,</span>
<span id="cb49-6"><a href="#cb49-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">x =</span> <span class="st">"Clusters"</span>, <span class="at">y =</span> <span class="st">"Feature count (n)"</span>)</span>
<span id="cb49-7"><a href="#cb49-7" aria-hidden="true" tabindex="-1"></a>plot2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(<span class="fu">rowData</span>(tse)), <span class="fu">aes</span>(<span class="at">x =</span> clusters_kendall)) <span class="sc">+</span></span>
<span id="cb49-8"><a href="#cb49-8" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_bar</span>() <span class="sc">+</span></span>
<span id="cb49-9"><a href="#cb49-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">title =</span> <span class="st">"CAG size distribution (1 - tau)"</span>,</span>
<span id="cb49-10"><a href="#cb49-10" aria-hidden="true" tabindex="-1"></a>         <span class="at">x =</span> <span class="st">"Clusters"</span>, <span class="at">y =</span> <span class="st">"Feature count (n)"</span>)</span>
<span id="cb49-11"><a href="#cb49-11" aria-hidden="true" tabindex="-1"></a>plot1 <span class="sc">+</span> plot2 <span class="sc">+</span> <span class="fu">plot_layout</span>(<span class="at">ncol =</span> <span class="dv">2</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="11_taxonomic_information_files/figure-html/taxa_clustering_histogram-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>It’s also possible to merge the rows by cluster.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/taxa_clustering_row_merge_8d420a086611c8055f6f932b356e59ae">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Aggregate clusters as a sum of each cluster values</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>tse_merged <span class="ot">&lt;-</span> <span class="fu">mergeFeatures</span>(tse, clusters_euclidean)</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>tse_merged</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>class: TreeSummarizedExperiment 
dim: 5 58 
metadata(0):
assays(3): counts clr z
rownames(5): 1 2 3 4 5
rowData names(8): kingdom phylum ... hclustEuclidean hclustKendall
colnames(58): ID1 ID2 ... ID57 ID58
colData names(5): Sample Geographical_location Gender Age Diet
reducedDimNames(0):
mainExpName: NULL
altExpNames(0):
rowLinks: NULL
rowTree: NULL
colLinks: NULL
colTree: NULL</code></pre>
</div>
</div>
<p>We can note that it worked as planned since there were 5 clusters and there are now 5 rows.</p>
</section>
</section>
<section id="assay-transform" class="level2" data-number="7.4">
<h2 data-number="7.4" class="anchored" data-anchor-id="assay-transform"><span class="header-section-number">7.4</span> Data transformation</h2>
<p>Data transformations are common in microbiome analysis. Examples include the logarithmic transformation, calculation of relative abundances (percentages), and compositionality-aware transformations such as the centered log-ratio transformation (clr).</p>
<p>In mia package, transformations are applied to abundance data. The transformed abundance table is stored back to ‘assays’. mia includes transformation function (‘transformAssay()’) which applies sample-wise or column-wise transformation when MARGIN = ‘samples’, feature-wise or row-wise transformation when MARGIN = ‘features’.</p>
<p>For a complete list of available transformations and parameters, see function <a href="https://microbiome.github.io/mia/reference/transformAssay.html">help</a>.</p>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-24_4e107509ea0fd7b98cce6c1fcf522130">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> tseGlobalPatterns</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">assay.type =</span> <span class="st">"counts"</span>, <span class="at">method =</span> <span class="st">"relabundance"</span>, <span class="at">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(<span class="at">x =</span> tse, <span class="at">assay.type =</span> <span class="st">"relabundance"</span>, <span class="at">method =</span> <span class="st">"clr"</span>, </span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>                      <span class="at">pseudocount =</span> <span class="dv">1</span>, <span class="at">name =</span> <span class="st">"clr"</span>)</span>
<span id="cb52-5"><a href="#cb52-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-6"><a href="#cb52-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">assay</span>(tse, <span class="st">"clr"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                        CL3        CC1        SV1    M31Fcsw
Class:Thermoprotei               -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
Class:Thermoprotei               -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
Species:Sulfolobusacidocaldarius -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
Class:Sd-NA                      -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
Class:Sd-NA                      -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
Class:Sd-NA                      -5.078e-05 -5.105e-05 -5.055e-05 -4.975e-05
                                    M11Fcsw    M31Plmr    M11Plmr    F21Plmr
Class:Thermoprotei               -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
Class:Thermoprotei               -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
Species:Sulfolobusacidocaldarius -4.947e-05 -4.931e-05 -4.658e-05 -4.671e-05
Class:Sd-NA                      -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
Class:Sd-NA                      -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
Class:Sd-NA                      -4.947e-05 -4.931e-05 -4.879e-05 -4.671e-05
                                    M31Tong    M11Tong   LMEpi24M   SLEpi20M
Class:Thermoprotei               -4.846e-05 -4.257e-05 -4.756e-05 -4.837e-05
Class:Thermoprotei               -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
Species:Sulfolobusacidocaldarius -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
Class:Sd-NA                      -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
Class:Sd-NA                      -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
Class:Sd-NA                      -4.846e-05 -4.257e-05 -4.756e-05 -4.918e-05
                                     AQC1cm     AQC4cm     AQC7cm        NP2
Class:Thermoprotei               -2.385e-05 -4.438e-06  2.787e-05 -4.731e-05
Class:Thermoprotei               -4.660e-05 -4.568e-05 -4.428e-05 -4.915e-05
Species:Sulfolobusacidocaldarius -4.660e-05 -4.652e-05 -4.777e-05 -4.915e-05
Class:Sd-NA                      -4.660e-05 -3.726e-05 -3.090e-05 -4.915e-05
Class:Sd-NA                      -4.660e-05 -4.568e-05 -4.719e-05 -4.915e-05
Class:Sd-NA                      -4.660e-05 -4.610e-05 -4.603e-05 -4.915e-05
                                        NP3        NP5    TRRsed1    TRRsed2
Class:Thermoprotei               -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
Class:Thermoprotei               -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
Species:Sulfolobusacidocaldarius -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
Class:Sd-NA                      -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
Class:Sd-NA                      -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
Class:Sd-NA                      -5.068e-05 -5.083e-05 -3.909e-05 -4.927e-05
                                    TRRsed3       TS28       TS29      Even1
Class:Thermoprotei               -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
Class:Thermoprotei               -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
Species:Sulfolobusacidocaldarius -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
Class:Sd-NA                      -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
Class:Sd-NA                      -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
Class:Sd-NA                      -4.829e-05 -5.016e-05 -4.934e-05 -5.046e-05
                                      Even2      Even3
Class:Thermoprotei               -5.017e-05 -5.034e-05
Class:Thermoprotei               -5.017e-05 -5.034e-05
Species:Sulfolobusacidocaldarius -5.017e-05 -5.034e-05
Class:Sd-NA                      -5.017e-05 -5.034e-05
Class:Sd-NA                      -5.017e-05 -5.034e-05
Class:Sd-NA                      -5.017e-05 -5.034e-05</code></pre>
</div>
</div>
<ul>
<li>In ‘pa’ transformation, abundance table is converted to present/absent table.</li>
</ul>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-25_5ff15ea3250c29f2646094412d05c244">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">method =</span> <span class="st">"pa"</span>)</span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">assay</span>(tse, <span class="st">"pa"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                                 CL3 CC1 SV1 M31Fcsw M11Fcsw M31Plmr M11Plmr
Class:Thermoprotei                 0   0   0       0       0       0       0
Class:Thermoprotei                 0   0   0       0       0       0       0
Species:Sulfolobusacidocaldarius   0   0   0       0       0       0       1
Class:Sd-NA                        0   0   0       0       0       0       0
Class:Sd-NA                        0   0   0       0       0       0       0
Class:Sd-NA                        0   0   0       0       0       0       0
                                 F21Plmr M31Tong M11Tong LMEpi24M SLEpi20M
Class:Thermoprotei                     0       0       0        0        1
Class:Thermoprotei                     0       0       0        0        0
Species:Sulfolobusacidocaldarius       0       0       0        0        0
Class:Sd-NA                            0       0       0        0        0
Class:Sd-NA                            0       0       0        0        0
Class:Sd-NA                            0       0       0        0        0
                                 AQC1cm AQC4cm AQC7cm NP2 NP3 NP5 TRRsed1
Class:Thermoprotei                    1      1      1   1   0   0       0
Class:Thermoprotei                    0      1      1   0   0   0       0
Species:Sulfolobusacidocaldarius      0      0      0   0   0   0       0
Class:Sd-NA                           0      1      1   0   0   0       0
Class:Sd-NA                           0      1      1   0   0   0       0
Class:Sd-NA                           0      1      1   0   0   0       0
                                 TRRsed2 TRRsed3 TS28 TS29 Even1 Even2 Even3
Class:Thermoprotei                     0       0    0    0     0     0     0
Class:Thermoprotei                     0       0    0    0     0     0     0
Species:Sulfolobusacidocaldarius       0       0    0    0     0     0     0
Class:Sd-NA                            0       0    0    0     0     0     0
Class:Sd-NA                            0       0    0    0     0     0     0
Class:Sd-NA                            0       0    0    0     0     0     0</code></pre>
</div>
</div>
<div class="cell" data-hash="11_taxonomic_information_cache/html/unnamed-chunk-26_bb6c465aef012fcd8e6fb96bc788a319">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># list of abundance tables that assays slot contains</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a><span class="fu">assays</span>(tse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of length 4
names(4): counts relabundance clr pa</code></pre>
</div>
</div>


<div id="refs" class="references csl-bib-body hanging-indent" role="list">
<div id="ref-Callahan2016dada2" class="csl-entry" role="listitem">
Callahan, Benjamin J, Paul J McMurdie, Michael J Rosen, Andrew W Han, Amy Jo A Johnson, and Susan P Holmes. 2016. <span>“DADA2: High-Resolution Sample Inference from Illumina Amplicon Data.”</span> <em>Nature Methods</em> 13: 581–83. <a href="https://doi.org/10.1038/nmeth.3869">https://doi.org/10.1038/nmeth.3869</a>.
</div>
<div id="ref-R_TreeSummarizedExperiment" class="csl-entry" role="listitem">
Huang, Ruizhu. 2020. <em>TreeSummarizedExperiment: A S4 Class for Data with Tree Structures</em>.
</div>
<div id="ref-R_DECIPHER" class="csl-entry" role="listitem">
Wright, Erik. 2020. <em>DECIPHER: Tools for Curating, Analyzing, and Manipulating Biological Sequences</em>.
</div>
</div>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./12_quality_control.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./14_alpha_diversity.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">Community Diversity</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>