<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>6&nbsp; Exploration and Quality Control</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./14_alpha_diversity.html" rel="next">
<link href="./10_manipulation.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>


</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./12_quality_control.html"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Orchestrating Microbiome Analysis with Bioconductor</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="" rel="" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-twitter"></i></a>
    <a href="https://microbiome.github.io/mia/" rel="" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-facebook"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Orchestrating Microbiome Analysis with Bioconductor</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12_quality_control.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">global knitr options</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./30_differential_abundance.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Developers</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Sessioninfo</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./99_bibliography.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bibliography</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#abundance" id="toc-abundance" class="nav-link active" data-scroll-target="#abundance"><span class="header-section-number">6.1</span> Abundance</a></li>
  <li><a href="#prevalence" id="toc-prevalence" class="nav-link" data-scroll-target="#prevalence"><span class="header-section-number">6.2</span> Prevalence</a>
  <ul class="collapse">
  <li><a href="#prevalence-analysis" id="toc-prevalence-analysis" class="nav-link" data-scroll-target="#prevalence-analysis"><span class="header-section-number">6.2.1</span> Prevalence analysis</a></li>
  <li><a href="#rare-taxa" id="toc-rare-taxa" class="nav-link" data-scroll-target="#rare-taxa"><span class="header-section-number">6.2.2</span> Rare taxa</a></li>
  <li><a href="#plotting-prevalence" id="toc-plotting-prevalence" class="nav-link" data-scroll-target="#plotting-prevalence"><span class="header-section-number">6.2.3</span> Plotting prevalence</a></li>
  </ul></li>
  <li><a href="#qc" id="toc-qc" class="nav-link" data-scroll-target="#qc"><span class="header-section-number">6.3</span> Quality control</a>
  <ul class="collapse">
  <li><a href="#top-taxa" id="toc-top-taxa" class="nav-link" data-scroll-target="#top-taxa"><span class="header-section-number">6.3.1</span> Top taxa</a></li>
  <li><a href="#library-size-read-count" id="toc-library-size-read-count" class="nav-link" data-scroll-target="#library-size-read-count"><span class="header-section-number">6.3.2</span> Library size / read count</a></li>
  <li><a href="#contaminant-sequences" id="toc-contaminant-sequences" class="nav-link" data-scroll-target="#contaminant-sequences"><span class="header-section-number">6.3.3</span> Contaminant sequences</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="quality-control" class="quarto-section-identifier"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<p>This chapter focuses on the quality control and exploration of microbiome data and establishes commonly used descriptive summaries. Familiarizing with the peculiarities of a given dataset is the essential basis for any data analysis and model building.</p>
<p>The dataset should not suffer from severe technical biases, and you should at least be aware of potential challenges, such as outliers, biases, unexpected patterns and so forth. Standard summaries and visualizations can help, and the rest comes with experience. The exploration and quality control can be iterative processes.</p>
<div class="cell" data-hash="12_quality_control_cache/html/unnamed-chunk-2_29d98f28c824d256d2d64b04296a9932">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="abundance" class="level2" data-number="6.1">
<h2 data-number="6.1" class="anchored" data-anchor-id="abundance"><span class="header-section-number">6.1</span> Abundance</h2>
<p>Abundance visualization is an important data exploration approach. <code>miaViz</code> offers the function <code>plotAbundanceDensity</code> to plot the most abundant taxa with several options.</p>
<p>Next, a few demonstrations are shown, using the <span class="citation" data-cites="Lahti2014">[@Lahti2014]</span> dataset. A Jitter plot based on relative abundance data, similar to the one presented at <span class="citation" data-cites="Salosensaari2021">[@Salosensaari2021]</span> supplementary figure 1, could be visualized as follows:</p>
<div class="cell" data-hash="12_quality_control_cache/html/unnamed-chunk-3_0006f2070931bdee2b569847e2fa8e7d">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load example data</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(miaTime)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(miaViz)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(hitchip1006)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> hitchip1006</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Add relative abundances</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse, <span class="at">MARGIN =</span> <span class="st">"samples"</span>, <span class="at">method =</span> <span class="st">"relabundance"</span>)</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Use argument names</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a><span class="co"># assay.type / assay.type / assay.type</span></span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a><span class="co"># depending on the mia package version</span></span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAbundanceDensity</span>(tse, <span class="at">layout =</span> <span class="st">"jitter"</span>, <span class="at">assay.type =</span> <span class="st">"relabundance"</span>,</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>                     <span class="at">n =</span> <span class="dv">40</span>, <span class="at">point_size=</span><span class="dv">1</span>, <span class="at">point_shape=</span><span class="dv">19</span>, <span class="at">point_alpha=</span><span class="fl">0.1</span>) <span class="sc">+</span> </span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>                     <span class="fu">scale_x_log10</span>(<span class="at">label=</span>scales<span class="sc">::</span>percent)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="12_quality_control_files/figure-html/unnamed-chunk-3-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The relative abundance values for the top-5 taxonomic features can be visualized as a density plot over a log scaled axis, with “nationality” indicated by colors:</p>
<div class="cell" data-hash="12_quality_control_cache/html/unnamed-chunk-4_b30c0825a34594da8a59058216e51cc0">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotAbundanceDensity</span>(tse, <span class="at">layout =</span> <span class="st">"density"</span>, <span class="at">assay.type =</span> <span class="st">"relabundance"</span>,</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>                     <span class="at">n =</span> <span class="dv">5</span>, <span class="at">colour_by=</span><span class="st">"nationality"</span>, <span class="at">point_alpha=</span><span class="dv">1</span><span class="sc">/</span><span class="dv">10</span>) <span class="sc">+</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_log10</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="12_quality_control_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="prevalence" class="level2" data-number="6.2">
<h2 data-number="6.2" class="anchored" data-anchor-id="prevalence"><span class="header-section-number">6.2</span> Prevalence</h2>
<p>Prevalence quantifies the frequency of samples where certain microbes were detected (above a given detection threshold). The prevalence can be given as sample size (N) or percentage (unit interval).</p>
<p>Investigating prevalence allows you either to focus on changes which pertain to the majority of the samples, or identify rare microbes, which may be <em>conditionally abundant</em> in a small number of samples.</p>
<p>The population prevalence (frequency) at a 1% relative abundance threshold (<code>detection = 1/100</code> and <code>as_relative = TRUE</code>), can look like this.</p>
<div class="cell" data-hash="12_quality_control_cache/html/exploration-prevalence_d4d9983bb3acfed3aad4fb11610a487c">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getPrevalence</span>(tse, <span class="at">detection =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">100</span>, <span class="at">sort =</span> <span class="cn">TRUE</span>, <span class="at">as_relative =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Faecalibacterium prausnitzii et rel.           Ruminococcus obeum et rel. 
                              0.9522                               0.9140 
  Oscillospira guillermondii et rel.        Clostridium symbiosum et rel. 
                              0.8801                               0.8714 
    Subdoligranulum variable at rel.     Clostridium orbiscindens et rel. 
                              0.8358                               0.8315 </code></pre>
</div>
</div>
<p>The function arguments <code>detection</code> and <code>as_relative</code> can also be used to access, how many samples do pass a threshold for raw counts. Here, the population prevalence (frequency) at the absolute abundance threshold (<code>as_relative = FALSE</code>) at read count 1 (<code>detection = 1</code>) is accessed.</p>
<div class="cell" data-hash="12_quality_control_cache/html/concepts_prevalence2_17de52a42a1d7b2c6f964ebe5792675e">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getPrevalence</span>(tse, <span class="at">detection =</span> <span class="dv">1</span>, <span class="at">sort =</span> <span class="cn">TRUE</span>, <span class="at">assay.type =</span> <span class="st">"counts"</span>,</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">as_relative =</span> <span class="cn">FALSE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>           Uncultured Mollicutes      Uncultured Clostridiales II 
                               1                                1 
      Uncultured Clostridiales I               Tannerella et rel. 
                               1                                1 
  Sutterella wadsworthia et rel. Subdoligranulum variable at rel. 
                               1                                1 </code></pre>
</div>
</div>
<p>If the output should be used for subsetting or storing the data in the <code>rowData</code>, set <code>sort = FALSE</code>.</p>
<section id="prevalence-analysis" class="level3" data-number="6.2.1">
<h3 data-number="6.2.1" class="anchored" data-anchor-id="prevalence-analysis"><span class="header-section-number">6.2.1</span> Prevalence analysis</h3>
<p>To investigate microbiome prevalence at a selected taxonomic level, two approaches are available.</p>
<p>First the data can be agglomerated to the taxonomic level and <code>getPrevalence</code> applied on the resulting object.</p>
<div class="cell" data-hash="12_quality_control_cache/html/unnamed-chunk-7_8fe67295e6a075b27d9bb578a9e47ef5">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Agglomerate taxa abundances to Phylum level, and add the new table</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="co"># to the altExp slot</span></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">altExp</span>(tse,<span class="st">"Phylum"</span>) <span class="ot">&lt;-</span> <span class="fu">mergeFeaturesByRank</span>(tse, <span class="st">"Phylum"</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Check prevalence for the Phylum abundance table from the altExp slot</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getPrevalence</span>(<span class="fu">altExp</span>(tse,<span class="st">"Phylum"</span>), <span class="at">detection =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">100</span>, <span class="at">sort =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">assay.type =</span> <span class="st">"counts"</span>, <span class="at">as_relative =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Firmicutes   Bacteroidetes  Actinobacteria  Proteobacteria Verrucomicrobia 
      1.0000000       0.9852302       0.4821894       0.2988705       0.1277150 
  Cyanobacteria 
      0.0008688 </code></pre>
</div>
</div>
<p>Alternatively, the <code>rank</code> argument could be set to perform the agglomeration on the fly.</p>
<div class="cell" data-hash="12_quality_control_cache/html/unnamed-chunk-8_16d449f0c3ee5faddf8b5b395be4f1f6">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">getPrevalence</span>(tse, <span class="at">rank =</span> <span class="st">"Phylum"</span>, <span class="at">detection =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">100</span>, <span class="at">sort =</span> <span class="cn">TRUE</span>,</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>                   <span class="at">assay.type =</span> <span class="st">"counts"</span>, <span class="at">as_relative =</span> <span class="cn">TRUE</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>     Firmicutes   Bacteroidetes  Actinobacteria  Proteobacteria Verrucomicrobia 
      1.0000000       0.9852302       0.4821894       0.2988705       0.1277150 
  Cyanobacteria 
      0.0008688 </code></pre>
</div>
</div>
<p>Note that, by default, <code>na.rm = TRUE</code> is used for agglomeration in <code>getPrevalence</code>, whereas the default for <code>mergeFeaturesByRank</code> is <code>FALSE</code> to prevent accidental data loss.</p>
<p>If you only need the names of the prevalent taxa, <code>getPrevalentFeatures</code> is available. This returns the taxa that exceed the given prevalence and detection thresholds.</p>
<div class="cell" data-hash="12_quality_control_cache/html/core-members_c926a31d8808f3927d834f3a808a02e3">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">getPrevalentFeatures</span>(tse, <span class="at">detection =</span> <span class="dv">0</span>, <span class="at">prevalence =</span> <span class="dv">50</span><span class="sc">/</span><span class="dv">100</span>)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>prev <span class="ot">&lt;-</span> <span class="fu">getPrevalentFeatures</span>(tse, <span class="at">detection =</span> <span class="dv">0</span>, <span class="at">prevalence =</span> <span class="dv">50</span><span class="sc">/</span><span class="dv">100</span>,</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">rank =</span> <span class="st">"Phylum"</span>, <span class="at">sort =</span> <span class="cn">TRUE</span>)</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>prev</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Note that the <code>detection</code> and <code>prevalence</code> thresholds are not the same, since <code>detection</code> can be applied to relative counts or absolute counts depending on whether <code>as_relative</code> is set <code>TRUE</code> or <code>FALSE</code></p>
<p>The function ‘getPrevalentAbundance’ can be used to check the total relative abundance of the prevalent taxa (between 0 and 1).</p>
</section>
<section id="rare-taxa" class="level3" data-number="6.2.2">
<h3 data-number="6.2.2" class="anchored" data-anchor-id="rare-taxa"><span class="header-section-number">6.2.2</span> Rare taxa</h3>
<p>Related functions are available for the analysis of rare taxa (<code>rareMembers</code>; <code>rareAbundance</code>; <code>lowAbundance</code>, <code>getRareFeatures</code>, <code>subsetByRareFeatures</code>).</p>
</section>
<section id="plotting-prevalence" class="level3" data-number="6.2.3">
<h3 data-number="6.2.3" class="anchored" data-anchor-id="plotting-prevalence"><span class="header-section-number">6.2.3</span> Plotting prevalence</h3>
<p>To plot the prevalence, add the prevalence of each taxon to <code>rowData</code>. Here, we are analysing the Phylum level abundances, which are stored in the <code>altExp</code> slot.</p>
<div class="cell" data-hash="12_quality_control_cache/html/unnamed-chunk-10_f24682cf62ccf98bcdfc29885b12afdf">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(<span class="fu">altExp</span>(tse,<span class="st">"Phylum"</span>))<span class="sc">$</span>prevalence <span class="ot">&lt;-</span> </span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">getPrevalence</span>(<span class="fu">altExp</span>(tse,<span class="st">"Phylum"</span>), <span class="at">detection =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">100</span>, <span class="at">sort =</span> <span class="cn">FALSE</span>,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>                  <span class="at">assay.type =</span> <span class="st">"counts"</span>, <span class="at">as_relative =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The prevalences can then be plotted using the plotting functions from the <code>scater</code> package.</p>
<div class="cell" data-hash="12_quality_control_cache/html/unnamed-chunk-11_6893facc68fdf0213da07fb05d7043a4">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRowData</span>(<span class="fu">altExp</span>(tse,<span class="st">"Phylum"</span>), <span class="st">"prevalence"</span>, <span class="at">colour_by =</span> <span class="st">"Phylum"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="12_quality_control_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The prevalence can also be visualized on the taxonomic tree with the <code>miaViz</code> package.</p>
<div class="cell" data-hash="12_quality_control_cache/html/unnamed-chunk-12_de968e97b44a863129eb1abba3c7bbe3">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">altExps</span>(tse) <span class="ot">&lt;-</span> <span class="fu">splitByRanks</span>(tse)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">altExps</span>(tse) <span class="ot">&lt;-</span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>   <span class="fu">lapply</span>(<span class="fu">altExps</span>(tse),</span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a>          <span class="cf">function</span>(y){</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>              <span class="fu">rowData</span>(y)<span class="sc">$</span>prevalence <span class="ot">&lt;-</span> </span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>                  <span class="fu">getPrevalence</span>(y, <span class="at">detection =</span> <span class="dv">1</span><span class="sc">/</span><span class="dv">100</span>, <span class="at">sort =</span> <span class="cn">FALSE</span>,</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>                                <span class="at">assay.type =</span> <span class="st">"counts"</span>, <span class="at">as_relative =</span> <span class="cn">TRUE</span>)</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>              y</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>          })</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>top_phyla <span class="ot">&lt;-</span> <span class="fu">getTopFeatures</span>(<span class="fu">altExp</span>(tse,<span class="st">"Phylum"</span>),</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>                        <span class="at">method=</span><span class="st">"prevalence"</span>,</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>                        <span class="at">top=</span>5L,</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>                        <span class="at">assay.type=</span><span class="st">"counts"</span>)</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>top_phyla_mean <span class="ot">&lt;-</span> <span class="fu">getTopFeatures</span>(<span class="fu">altExp</span>(tse,<span class="st">"Phylum"</span>),</span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>                             <span class="at">method=</span><span class="st">"mean"</span>,</span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>                             <span class="at">top=</span>5L,</span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>                             <span class="at">assay.type=</span><span class="st">"counts"</span>)</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">unsplitByRanks</span>(tse, <span class="at">ranks =</span> <span class="fu">taxonomyRanks</span>(tse)[<span class="dv">1</span><span class="sc">:</span><span class="dv">6</span>])</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">addTaxonomyTree</span>(x)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>After some preparation, the data is assembled and can be plotted with <code>plotRowTree</code>.</p>
<div class="cell" data-hash="12_quality_control_cache/html/plot-prev-prev_6ee85c56e593996af9110e106b829e1e">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(miaViz)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRowTree</span>(x[<span class="fu">rowData</span>(x)<span class="sc">$</span>Phylum <span class="sc">%in%</span> top_phyla,],</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">edge_colour_by =</span> <span class="st">"Phylum"</span>,</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">tip_colour_by =</span> <span class="st">"prevalence"</span>,</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_colour_by =</span> <span class="st">"prevalence"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12_quality_control_files/figure-html/plot-prev-prev-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Prevalence of top phyla as judged by prevalence</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-hash="12_quality_control_cache/html/plot-prev-mean_047d12e98c69b395ed4058997c65d22d">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotRowTree</span>(x[<span class="fu">rowData</span>(x)<span class="sc">$</span>Phylum <span class="sc">%in%</span> top_phyla_mean,],</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>            <span class="at">edge_colour_by =</span> <span class="st">"Phylum"</span>,</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>            <span class="at">tip_colour_by =</span> <span class="st">"prevalence"</span>,</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>            <span class="at">node_colour_by =</span> <span class="st">"prevalence"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12_quality_control_files/figure-html/plot-prev-mean-1.png" class="img-fluid figure-img" width="672"></p>
<figcaption class="figure-caption">Prevalence of top phyla as judged by mean abundance</figcaption>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="qc" class="level2" data-number="6.3">
<h2 data-number="6.3" class="anchored" data-anchor-id="qc"><span class="header-section-number">6.3</span> Quality control</h2>
<p>Next, let us load the <code>GlobalPatterns</code> dataset to illustrate standard microbiome data summaries.</p>
<div class="cell" data-hash="12_quality_control_cache/html/unnamed-chunk-15_ee57fba5398762a32e505c365fb3f63c">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"GlobalPatterns"</span>, <span class="at">package=</span><span class="st">"mia"</span>)</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> GlobalPatterns </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="top-taxa" class="level3" data-number="6.3.1">
<h3 data-number="6.3.1" class="anchored" data-anchor-id="top-taxa"><span class="header-section-number">6.3.1</span> Top taxa</h3>
<p>The <code>getTopFeatures</code> identifies top taxa in the data.</p>
<div class="cell" data-hash="12_quality_control_cache/html/top-feature-taxo_ed8cb6c9a5b9967c03e7c5f58808edce">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Pick the top taxa</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>top_features <span class="ot">&lt;-</span> <span class="fu">getTopFeatures</span>(tse, <span class="at">method=</span><span class="st">"median"</span>, <span class="at">top=</span><span class="dv">10</span>)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Check the information for these</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="fu">rowData</span>(tse)[top_features, <span class="fu">taxonomyRanks</span>(tse)]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 10 rows and 7 columns
           Kingdom         Phylum               Class             Order
       &lt;character&gt;    &lt;character&gt;         &lt;character&gt;       &lt;character&gt;
549656    Bacteria  Cyanobacteria         Chloroplast     Stramenopiles
331820    Bacteria  Bacteroidetes         Bacteroidia     Bacteroidales
317182    Bacteria  Cyanobacteria         Chloroplast     Stramenopiles
94166     Bacteria Proteobacteria Gammaproteobacteria    Pasteurellales
279599    Bacteria  Cyanobacteria    Nostocophycideae        Nostocales
158660    Bacteria  Bacteroidetes         Bacteroidia     Bacteroidales
329744    Bacteria Actinobacteria      Actinobacteria   Actinomycetales
326977    Bacteria Actinobacteria      Actinobacteria Bifidobacteriales
248140    Bacteria  Bacteroidetes         Bacteroidia     Bacteroidales
550960    Bacteria Proteobacteria Gammaproteobacteria Enterobacteriales
                   Family           Genus                Species
              &lt;character&gt;     &lt;character&gt;            &lt;character&gt;
549656                 NA              NA                     NA
331820     Bacteroidaceae     Bacteroides                     NA
317182                 NA              NA                     NA
94166     Pasteurellaceae     Haemophilus Haemophilusparainflu..
279599        Nostocaceae  Dolichospermum                     NA
158660     Bacteroidaceae     Bacteroides                     NA
329744             ACK-M1              NA                     NA
326977 Bifidobacteriaceae Bifidobacterium Bifidobacteriumadole..
248140     Bacteroidaceae     Bacteroides      Bacteroidescaccae
550960 Enterobacteriaceae     Providencia                     NA</code></pre>
</div>
</div>
</section>
<section id="library-size-read-count" class="level3" data-number="6.3.2">
<h3 data-number="6.3.2" class="anchored" data-anchor-id="library-size-read-count"><span class="header-section-number">6.3.2</span> Library size / read count</h3>
<p>The total counts/sample can be calculated using <code>perCellQCMetrics</code>/<code>addPerCellQC</code> from the <code>scater</code> package. The former one just calculates the values, whereas the latter one directly adds them to <code>colData</code>.</p>
<div class="cell" data-hash="12_quality_control_cache/html/lib-size_c252fd98193bce2415bb54e24b86373f">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scater)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a><span class="fu">perCellQCMetrics</span>(tse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 26 rows and 3 columns
              sum  detected     total
        &lt;numeric&gt; &lt;numeric&gt; &lt;numeric&gt;
CL3        864077      6964    864077
CC1       1135457      7679   1135457
SV1        697509      5729    697509
M31Fcsw   1543451      2667   1543451
M11Fcsw   2076476      2574   2076476
...           ...       ...       ...
TS28       937466      2679    937466
TS29      1211071      2629   1211071
Even1     1216137      4213   1216137
Even2      971073      3130    971073
Even3     1078241      2776   1078241</code></pre>
</div>
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">addPerCellQC</span>(tse)</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(tse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>DataFrame with 26 rows and 10 columns
        X.SampleID   Primer Final_Barcode Barcode_truncated_plus_T
          &lt;factor&gt; &lt;factor&gt;      &lt;factor&gt;                 &lt;factor&gt;
CL3        CL3      ILBC_01        AACGCA                   TGCGTT
CC1        CC1      ILBC_02        AACTCG                   CGAGTT
SV1        SV1      ILBC_03        AACTGT                   ACAGTT
M31Fcsw    M31Fcsw  ILBC_04        AAGAGA                   TCTCTT
M11Fcsw    M11Fcsw  ILBC_05        AAGCTG                   CAGCTT
...            ...      ...           ...                      ...
TS28         TS28   ILBC_25        ACCAGA                   TCTGGT
TS29         TS29   ILBC_26        ACCAGC                   GCTGGT
Even1        Even1  ILBC_27        ACCGCA                   TGCGGT
Even2        Even2  ILBC_28        ACCTCG                   CGAGGT
Even3        Even3  ILBC_29        ACCTGT                   ACAGGT
        Barcode_full_length SampleType
                   &lt;factor&gt;   &lt;factor&gt;
CL3             CTAGCGTGCGT      Soil 
CC1             CATCGACGAGT      Soil 
SV1             GTACGCACAGT      Soil 
M31Fcsw         TCGACATCTCT      Feces
M11Fcsw         CGACTGCAGCT      Feces
...                     ...        ...
TS28            GCATCGTCTGG      Feces
TS29            CTAGTCGCTGG      Feces
Even1           TGACTCTGCGG      Mock 
Even2           TCTGATCGAGG      Mock 
Even3           AGAGAGACAGG      Mock 
                                       Description       sum  detected
                                          &lt;factor&gt; &lt;numeric&gt; &lt;numeric&gt;
CL3     Calhoun South Carolina Pine soil, pH 4.9      864077      6964
CC1     Cedar Creek Minnesota, grassland, pH 6.1     1135457      7679
SV1     Sevilleta new Mexico, desert scrub, pH 8.3    697509      5729
M31Fcsw M3, Day 1, fecal swab, whole body study      1543451      2667
M11Fcsw M1, Day 1, fecal swab, whole body study      2076476      2574
...                                            ...       ...       ...
TS28                                       Twin #1    937466      2679
TS29                                       Twin #2   1211071      2629
Even1                                      Even1     1216137      4213
Even2                                      Even2      971073      3130
Even3                                      Even3     1078241      2776
            total
        &lt;numeric&gt;
CL3        864077
CC1       1135457
SV1        697509
M31Fcsw   1543451
M11Fcsw   2076476
...           ...
TS28       937466
TS29      1211071
Even1     1216137
Even2      971073
Even3     1078241</code></pre>
</div>
</div>
<p>The distribution of calculated library sizes can be visualized as a histogram (left), or by sorting the samples by library size (right).</p>
<div class="cell" data-hash="12_quality_control_cache/html/plot-viz-lib-size-1_9205ccd0fdf9459d45b64018252d37ed">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(<span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse))) <span class="sc">+</span></span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(<span class="at">x =</span> sum), <span class="at">color =</span> <span class="st">"black"</span>, <span class="at">fill =</span> <span class="st">"gray"</span>, <span class="at">bins =</span> <span class="dv">30</span>) <span class="sc">+</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>        <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Library size"</span>, <span class="at">y =</span> <span class="st">"Frequency (n)"</span>) <span class="sc">+</span> </span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>        <span class="co"># scale_x_log10(breaks = scales::trans_breaks("log10", function(x) 10^x), </span></span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>        <span class="co"># labels = scales::trans_format("log10", scales::math_format(10^.x))) +</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(), <span class="co"># Removes the grid</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>          <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>          <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>          <span class="at">panel.background =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true" tabindex="-1"></a>          <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">colour =</span> <span class="st">"black"</span>)) <span class="co"># Adds y-axis</span></span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(dplyr)</span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse)) <span class="sc">%&gt;%</span></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true" tabindex="-1"></a>        <span class="fu">arrange</span>(sum) <span class="sc">%&gt;%</span></span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true" tabindex="-1"></a>        <span class="fu">mutate</span>(<span class="at">index =</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">n</span>())</span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>(df, <span class="fu">aes</span>(<span class="at">y =</span> index, <span class="at">x =</span> sum<span class="sc">/</span><span class="fl">1e6</span>)) <span class="sc">+</span></span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true" tabindex="-1"></a>        <span class="fu">geom_point</span>() <span class="sc">+</span>  </span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true" tabindex="-1"></a>        <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"Library size (million reads)"</span>, <span class="at">y =</span> <span class="st">"Sample index"</span>) <span class="sc">+</span>  </span>
<span id="cb25-22"><a href="#cb25-22" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme_bw</span>() <span class="sc">+</span></span>
<span id="cb25-23"><a href="#cb25-23" aria-hidden="true" tabindex="-1"></a>        <span class="fu">theme</span>(<span class="at">panel.grid.major =</span> <span class="fu">element_blank</span>(), <span class="co"># Removes the grid</span></span>
<span id="cb25-24"><a href="#cb25-24" aria-hidden="true" tabindex="-1"></a>          <span class="at">panel.grid.minor =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb25-25"><a href="#cb25-25" aria-hidden="true" tabindex="-1"></a>          <span class="at">panel.border =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb25-26"><a href="#cb25-26" aria-hidden="true" tabindex="-1"></a>          <span class="at">panel.background =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb25-27"><a href="#cb25-27" aria-hidden="true" tabindex="-1"></a>          <span class="at">axis.line =</span> <span class="fu">element_line</span>(<span class="at">colour =</span> <span class="st">"black"</span>)) <span class="co"># Adds y-axis</span></span>
<span id="cb25-28"><a href="#cb25-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb25-29"><a href="#cb25-29" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(patchwork)</span>
<span id="cb25-30"><a href="#cb25-30" aria-hidden="true" tabindex="-1"></a>p1 <span class="sc">+</span> p2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12_quality_control_files/figure-html/plot-viz-lib-size-1-1.png" class="img-fluid figure-img" width="768"></p>
<figcaption class="figure-caption">Library size distribution.</figcaption>
</figure>
</div>
</div>
</div>
<p>Library sizes other variables from <code>colData</code> can be visualized by using specified function called <code>plotColData</code>.</p>
<div class="cell" data-hash="12_quality_control_cache/html/plot-viz-lib-size-2_064cee0a13b00e9e91d37efa1dc73124">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Sort samples by read count, order the factor levels, and store back to tse as DataFrame</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a><span class="co"># </span><span class="al">TODO</span><span class="co">: plotColData could include an option for sorting samples based on colData variables</span></span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(tse) <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse)) <span class="sc">%&gt;%</span></span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a>                 <span class="fu">arrange</span>(X.SampleID) <span class="sc">%&gt;%</span></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a>             <span class="fu">mutate</span>(<span class="at">X.SampleID =</span> <span class="fu">factor</span>(X.SampleID, <span class="at">levels=</span>X.SampleID)) <span class="sc">%&gt;%</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>         DataFrame</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(tse,<span class="st">"sum"</span>,<span class="st">"X.SampleID"</span>, <span class="at">colour_by =</span> <span class="st">"SampleType"</span>) <span class="sc">+</span> </span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust=</span><span class="dv">1</span>)) <span class="sc">+</span></span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">y =</span> <span class="st">"Library size (N)"</span>, <span class="at">x =</span> <span class="st">"Sample ID"</span>)       </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12_quality_control_files/figure-html/plot-viz-lib-size-2-1.png" class="img-fluid figure-img" width="768"></p>
<figcaption class="figure-caption">Library sizes per sample.</figcaption>
</figure>
</div>
</div>
</div>
<div class="cell" data-hash="12_quality_control_cache/html/plot-viz-lib-size-3_0530cc366639fefdc96c88b259d6ab3e">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plotColData</span>(tse,<span class="st">"sum"</span>,<span class="st">"SampleType"</span>, <span class="at">colour_by =</span> <span class="st">"SampleType"</span>) <span class="sc">+</span> </span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust=</span><span class="dv">1</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="12_quality_control_files/figure-html/plot-viz-lib-size-3-1.png" class="img-fluid figure-img" width="768"></p>
<figcaption class="figure-caption">Library sizes per sample type.</figcaption>
</figure>
</div>
</div>
</div>
<p>In addition, data can be rarefied with <a href="https://microbiome.github.io/mia/reference/subsampleCounts.html">subsampleCounts</a>, which normalises the samples to an equal number of reads. However, this practice has been discouraged for the analysis of differentially abundant microorganisms (see <span class="citation" data-cites="mcmurdie2014waste">[@mcmurdie2014waste]</span>).</p>
</section>
<section id="contaminant-sequences" class="level3" data-number="6.3.3">
<h3 data-number="6.3.3" class="anchored" data-anchor-id="contaminant-sequences"><span class="header-section-number">6.3.3</span> Contaminant sequences</h3>
<p>Samples might be contaminated with exogenous sequences. The impact of each contaminant can be estimated based on their frequencies and concentrations across the samples.</p>
<p>The following <a href="https://microbiome.github.io/mia/reference/isContaminant.html">decontam functions</a> are based on the <span class="citation" data-cites="davis2018simple">[@davis2018simple]</span> and support such functionality:</p>
<ul>
<li><code>isContaminant</code>, <code>isNotContaminant</code></li>
<li><code>addContaminantQC</code>, <code>addNotContaminantQC</code></li>
</ul>


</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./10_manipulation.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./14_alpha_diversity.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Diversity</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>