<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.450">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>10&nbsp; Differential Abundance</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="site_libs/quarto-nav/quarto-nav.js"></script>
<script src="site_libs/quarto-nav/headroom.min.js"></script>
<script src="site_libs/clipboard/clipboard.min.js"></script>
<script src="site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="site_libs/quarto-search/fuse.min.js"></script>
<script src="site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="./">
<link href="./40_machine_learning.html" rel="next">
<link href="./21_microbiome_community.html" rel="prev">
<script src="site_libs/quarto-html/quarto.js"></script>
<script src="site_libs/quarto-html/popper.min.js"></script>
<script src="site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="site_libs/quarto-html/anchor.min.js"></script>
<link href="site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-text-highlighting-styles">
<link href="site_libs/quarto-html/quarto-syntax-highlighting-dark.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-text-highlighting-styles">
<script src="site_libs/bootstrap/bootstrap.min.js"></script>
<link href="site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" class="quarto-color-scheme" id="quarto-bootstrap" data-mode="light">
<link href="site_libs/bootstrap/bootstrap-dark.min.css" rel="prefetch" class="quarto-color-scheme quarto-color-alternate" id="quarto-bootstrap" data-mode="dark">
<script id="quarto-search-options" type="application/json">{
  "location": "sidebar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "start",
  "type": "textbox",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit",
    "search-label": "Search"
  }
}</script>

  <script src="https://polyfill.io/v3/polyfill.min.js?features=es6"></script>
  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="nav-sidebar floating">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="./30_differential_abundance.html"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
      <button type="button" class="btn quarto-search-button" aria-label="" onclick="window.quartoOpenSearch();">
        <i class="bi bi-search"></i>
      </button>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation floating overflow-auto">
    <div class="pt-lg-2 mt-2 text-left sidebar-header">
    <div class="sidebar-title mb-0 py-0">
      <a href="./">Orchestrating Microbiome Analysis with Bioconductor</a> 
        <div class="sidebar-tools-main tools-wide">
    <a href="" rel="" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-twitter"></i></a>
    <a href="https://microbiome.github.io/mia/" rel="" title="" class="quarto-navigation-tool px-1" aria-label=""><i class="bi bi-facebook"></i></a>
  <a href="" class="quarto-color-scheme-toggle quarto-navigation-tool  px-1" onclick="window.quartoToggleColorScheme(); return false;" title="Toggle dark mode"><i class="bi"></i></a>
</div>
    </div>
      </div>
        <div class="mt-2 flex-shrink-0 align-items-center">
        <div class="sidebar-search">
        <div id="quarto-search" class="" title="Search"></div>
        </div>
        </div>
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./index.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">1</span>&nbsp; <span class="chapter-title">Orchestrating Microbiome Analysis with Bioconductor</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./01_intro.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">2</span>&nbsp; <span class="chapter-title">Introduction</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./06_packages.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">3</span>&nbsp; <span class="chapter-title">Packages</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./04_containers.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">4</span>&nbsp; <span class="chapter-title">Microbiome Data</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./10_manipulation.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">5</span>&nbsp; <span class="chapter-title">Data Manipulation</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./12_quality_control.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">6</span>&nbsp; <span class="chapter-title">Exploration and Quality Control</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./14_alpha_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">7</span>&nbsp; <span class="chapter-title">Community Diversity</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./20_beta_diversity.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">8</span>&nbsp; <span class="chapter-title">global knitr options</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./21_microbiome_community.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Composition</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./30_differential_abundance.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./40_machine_learning.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Machine Learning</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./23_multi-assay_analyses.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">12</span>&nbsp; <span class="chapter-title">Multi-Assay Analyses</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./19_visualization_techniques.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">13</span>&nbsp; <span class="chapter-title">Visualization</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./80_training.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">14</span>&nbsp; <span class="chapter-title">Training</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./95_resources.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">15</span>&nbsp; <span class="chapter-title">Resources</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./98_exercises.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">16</span>&nbsp; <span class="chapter-title">Exercises</span></span></a>
  </div>
</li>
        <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./97_extra_materials.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text"><span class="chapter-number">17</span>&nbsp; <span class="chapter-title">Extra material</span></span></a>
  </div>
</li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true">
 <span class="menu-text">Appendices</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./90_acknowledgments.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Developers</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./Session_info.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Sessioninfo</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="./99_bibliography.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Bibliography</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#statistical-challenges-of-microbiome-data" id="toc-statistical-challenges-of-microbiome-data" class="nav-link active" data-scroll-target="#statistical-challenges-of-microbiome-data"><span class="header-section-number">10.1</span> Statistical challenges of microbiome data</a></li>
  <li><a href="#using-the-tools" id="toc-using-the-tools" class="nav-link" data-scroll-target="#using-the-tools"><span class="header-section-number">10.2</span> Using the tools</a>
  <ul class="collapse">
  <li><a href="#preparing-the-data-for-daa" id="toc-preparing-the-data-for-daa" class="nav-link" data-scroll-target="#preparing-the-data-for-daa"><span class="header-section-number">10.2.1</span> Preparing the data for DAA</a></li>
  <li><a href="#aldex2" id="toc-aldex2" class="nav-link" data-scroll-target="#aldex2"><span class="header-section-number">10.2.2</span> ALDEx2</a></li>
  <li><a href="#ancom-bc" id="toc-ancom-bc" class="nav-link" data-scroll-target="#ancom-bc"><span class="header-section-number">10.2.3</span> ANCOM-BC</a></li>
  <li><a href="#maaslin2" id="toc-maaslin2" class="nav-link" data-scroll-target="#maaslin2"><span class="header-section-number">10.2.4</span> MaAsLin2</a></li>
  <li><a href="#linda" id="toc-linda" class="nav-link" data-scroll-target="#linda"><span class="header-section-number">10.2.5</span> LinDA</a></li>
  <li><a href="#zicoseq" id="toc-zicoseq" class="nav-link" data-scroll-target="#zicoseq"><span class="header-section-number">10.2.6</span> ZicoSeq</a></li>
  <li><a href="#philr" id="toc-philr" class="nav-link" data-scroll-target="#philr"><span class="header-section-number">10.2.7</span> PhILR</a></li>
  <li><a href="#comparison-of-methods" id="toc-comparison-of-methods" class="nav-link" data-scroll-target="#comparison-of-methods"><span class="header-section-number">10.2.8</span> Comparison of methods</a></li>
  </ul></li>
  <li><a href="#daa-with-confounding" id="toc-daa-with-confounding" class="nav-link" data-scroll-target="#daa-with-confounding"><span class="header-section-number">10.3</span> DAA with confounding</a>
  <ul class="collapse">
  <li><a href="#selecting-confounders" id="toc-selecting-confounders" class="nav-link" data-scroll-target="#selecting-confounders"><span class="header-section-number">10.3.1</span> Selecting confounders</a></li>
  <li><a href="#ancom-bc-1" id="toc-ancom-bc-1" class="nav-link" data-scroll-target="#ancom-bc-1"><span class="header-section-number">10.3.2</span> ANCOM-BC</a></li>
  <li><a href="#linda-1" id="toc-linda-1" class="nav-link" data-scroll-target="#linda-1"><span class="header-section-number">10.3.3</span> LinDA</a></li>
  <li><a href="#zicoseq-1" id="toc-zicoseq-1" class="nav-link" data-scroll-target="#zicoseq-1"><span class="header-section-number">10.3.4</span> ZicoSeq</a></li>
  </ul></li>
  <li><a href="#additional-resources" id="toc-additional-resources" class="nav-link" data-scroll-target="#additional-resources"><span class="header-section-number">10.4</span> Additional resources</a></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title"><span id="differential-abundance" class="quarto-section-identifier"><span class="chapter-number">10</span>&nbsp; <span class="chapter-title">Differential Abundance</span></span></h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  

</header>

<script>
document.addEventListener("click", function (event) {
    if (event.target.classList.contains("rebook-collapse")) {
        event.target.classList.toggle("active");
        var content = event.target.nextElementSibling;
        if (content.style.display === "block") {
            content.style.display = "none";
        } else {
            content.style.display = "block";
        }
    }
})
</script>
<style>
.rebook-collapse {
  background-color: #eee;
  color: #444;
  cursor: pointer;
  padding: 18px;
  width: 100%;
  border: none;
  text-align: left;
  outline: none;
  font-size: 15px;
}

.rebook-content {
  padding: 0 18px;
  display: none;
  overflow: hidden;
  background-color: #f1f1f1;
}
</style>
<p>This section provides an introduction to Differential Abundance Analysis (DAA), which is used to identify differences in the abundances of individual taxa (at any taxonomic level) between two or more groups, such as treatment and control. Here, we demonstrate its implementation on Tengeler2020, described in chapter @ref(tengeler-desc).</p>
<p>The goal of DAA is to identify biomarkers of a certain phenotype or condition, and gain understanding of a complex system by looking at its isolated components. For example, the identification of a bacterial taxon that is differentially abundant between healthy patients and diseased patients can lead to important insights into the pathophysiology of the disease. In other words, differentially abundant taxa can be involved in the dynamics of the disease, which in turn helps understand the system as a whole. Despite its relevance in current research, the DAA approach has also been subject to debate <span class="citation" data-cites="Quinn2021">[@Quinn2021]</span>.</p>
<section id="statistical-challenges-of-microbiome-data" class="level2" data-number="10.1">
<h2 data-number="10.1" class="anchored" data-anchor-id="statistical-challenges-of-microbiome-data"><span class="header-section-number">10.1</span> Statistical challenges of microbiome data</h2>
<p>Microbiome data display unique properties that are exclusively addressed by DAA tools developed for microbiome analysis. Specifically, microbiome data are characterized by high variability, zero-inflation and compositionality. <strong>High variability</strong> expresses that abundance of taxa often varies by several orders of magnitude from sample to sample. <strong>Zero-inflation</strong> means that typically more than 70% of the values are zeros, which could be due to either physical absence (structural zeros) or insufficient sampling effort (sampling zeros). <strong>Compositionality</strong> implies that a change in the absolute abundance of one taxon will lead to apparent variations in the relative abundances of other taxa in the same sample. If neglected, such properties may cause significant bias in the results of DAA. Therefore, several approaches have been developed to address the unique properties of microbiome data and provide statistically useful results.</p>
<p>The first approach to target zero-inflated data consists of specialized models, such as <strong>over-dispersed count models</strong> and <strong>zero-inflated mixture models</strong>. DESeq2, edgeR and corncorb are based on over-dispersed count models, whereas metagenomeSeq, RAIDA, ZIBB and Omnibus implement zero-inflated mixture models to address zero-inflation. Typically, these models assume a negative binomial, beta-binomial or normal/log-normal distribution. Alternatively, <strong>zero imputation</strong> also represents a valid approach to deal with zero-inflated data. ALDEx2 and eBay apply a Bayesian model to impute the zeros when working with proportion data, accounting for sampling variability and sequencing depth variation. Other methods, such as MaAsLin2 and ANCOMBC impute the zeros with a pseudo-count strategy.</p>
<p>Regarding the compositionality of microbiome data, several approaches have been developed to perform <strong>robust normalization</strong> with methods specifically designed to reduce the bias found in compositional data. Some examples include trimmed mean of M-values (TMM) normalization used by edgeR, relative log expression (RLE) normalization used by DESeq2, cumulative sum scaling (CSS) normalization used by metagenomeSeq, centered log-ratio transformation (CLR) normalization used by ALDEx2 and geometric mean of pairwise ratios (GMPR) normalization used by Omnibus and Wrench normalization <span class="citation" data-cites="Kumar2018">[@Kumar2018]</span>, which corrects the compositional bias by an empirical Bayes approach. Other methods to deal with compositional data entail reference taxa approach used by DACOMP and RAIDA, analyzing the pattern of pairwise log ratios as done by ANCOM and bias-correction applied by ANCOMBC.</p>
<p>We recommend to have a look at <span class="citation" data-cites="Nearing2022">@Nearing2022</span>. In this study, multiple DAA methods were applied to 38 different datasets and their results were compared to one another. Because each method follows a slightly different approach in terms of assumptions and normalization techniques, it was shown that results on the same dataset may differ substantially depending on the method. Recently, <span class="citation" data-cites="Yang2022">@Yang2022</span> comprehensively evaluated DAA methods via a semi-parametric framework and 106 real datasets. This study also concluded that different methods can produce contradictory results, creating the risk of cherry-picking the most favorable options for one’s own hypothesis. Therefore, it is highly recommended to perform DAA with multiple methods to determine whether the findings can be reproduced by different approaches. Built on the findings of <span class="citation" data-cites="Calgaro2020">@Calgaro2020</span>, the <em>benchdamic</em> <span class="citation" data-cites="Calgaro2022">[@Calgaro2022]</span> package could offer a valuable support in this regard. Through a comprehensive evaluation process it serves both practitioners by comparing DA methods from existing literature, and method developers by providing an impartial tool to evaluate their new approaches in comparison to what is already available. For details, check its extensive <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/benchdamic/inst/doc/intro.html">vignette</a>.</p>
</section>
<section id="using-the-tools" class="level2" data-number="10.2">
<h2 data-number="10.2" class="anchored" data-anchor-id="using-the-tools"><span class="header-section-number">10.2</span> Using the tools</h2>
<p>In this section we demonstrate the use of four methods that can be recommended based on recent literature (ANCOM-BC <span class="citation" data-cites="ancombc2020">[@ancombc2020]</span>, <em>ALDEx2</em> <span class="citation" data-cites="Gloor2016">[@Gloor2016]</span>, <em>Maaslin2</em> <span class="citation" data-cites="Mallick2020">[@Mallick2020]</span>, <em>LinDA</em> <span class="citation" data-cites="Zhou2022">[@Zhou2022]</span> and <em>ZicoSeq</em> <span class="citation" data-cites="Yang2022">[@Yang2022]</span>).</p>
<p>The purpose of this section is to show how to perform DAA in R, not how to correctly do causal inference. Depending on your experimental setup and your theory, you must determine how to specify any model exactly. E.g., there might be confounding factors that might drive (the absence of) differences between the shown groups that we ignore here for simplicity. Or your dataset is repeated sampling design, matched-pair design or the general longitudianl design. We will demonstrate how to include covariates in those models. We picked a dataset that merely has microbial abundances in a TSE object as well as a grouping variable in the sample data. We simplify the examples by only including two of the three groups.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/import-daa-data_9bbc8bb21ab309bc8736948f27f5587c">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(mia)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Import dataset</span></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">data</span>(<span class="st">"Tengeler2020"</span>, <span class="at">package =</span> <span class="st">"mia"</span>)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> Tengeler2020</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Show patient status by cohort</span></span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">table</span>(tse<span class="sc">$</span>patient_status, tse<span class="sc">$</span>cohort) <span class="sc">%&gt;%</span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">Cohort_1</th>
<th style="text-align: right;">Cohort_2</th>
<th style="text-align: right;">Cohort_3</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">ADHD</td>
<td style="text-align: right;">4</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">4</td>
</tr>
<tr class="even">
<td style="text-align: left;">Control</td>
<td style="text-align: right;">6</td>
<td style="text-align: right;">5</td>
<td style="text-align: right;">3</td>
</tr>
</tbody>
</table>
</div>
</div>
<section id="preparing-the-data-for-daa" class="level3" data-number="10.2.1">
<h3 data-number="10.2.1" class="anchored" data-anchor-id="preparing-the-data-for-daa"><span class="header-section-number">10.2.1</span> Preparing the data for DAA</h3>
<p>Before starting the analysis, it is recommended to reduce the size and complexity of the data to make the results more reproducible. For this purpose, we agglomerate the features by genus and filter them by a prevalence threshold of 10%.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/prep-daa-data_8ded5f31b1951411b0bec82496ea56db">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Agglomerate by genus and subset by prevalence</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">subsetByPrevalentTaxa</span>(tse,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>                             <span class="at">rank =</span> <span class="st">"Genus"</span>,</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>                             <span class="at">prevalence =</span> <span class="dv">10</span> <span class="sc">/</span> <span class="dv">100</span>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Transform count assay to relative abundances</span></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>tse <span class="ot">&lt;-</span> <span class="fu">transformAssay</span>(tse,</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>                      <span class="at">assay.type =</span> <span class="st">"counts"</span>,</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>                      <span class="at">method =</span> <span class="st">"relabundance"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>While some DAA tools provide optional arguments for prevalence filtering, here we filtered the tse object directly. This way, we ensure that the input data remains the same when multiple tools are used.</p>
</section>
<section id="aldex2" class="level3" data-number="10.2.2">
<h3 data-number="10.2.2" class="anchored" data-anchor-id="aldex2"><span class="header-section-number">10.2.2</span> ALDEx2</h3>
<p>In this section, we will show how to perform DAA with ALDEx2, which can be regarded as the method of choice for its consistency, as it normally identifies features that are also found by complementary methods <span class="citation" data-cites="Nearing2022">[@Nearing2022]</span>. A more extensive introduction to its functionality is available in the <a href="https://bioconductor.org/packages/release/bioc/vignettes/ALDEx2/inst/doc/ALDEx2_vignette.html">ALDEx2 vignette</a>.</p>
<p>ALDEx2 estimates technical variation within each sample per taxon by utilizing the Dirichlet distribution. It furthermore applies the CLR transformation (or closely related log-ratio transforms). Depending on the experimental setup, it will perform a two sample Welch’s t-test and Wilcoxon test or a one-way ANOVA and Kruskal-Wallis test. For more complex study designs, there is a possibility to utilize the <code>glm</code> functionality within ALDEx2. The Benjamini-Hochberg procedure is applied by default to correct for multiple testing.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/run-aldex2_c7f6ef8a670626881eae03a964d77a32">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load package</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ALDEx2)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate Monte Carlo samples of the Dirichlet distribution for each sample.</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert each instance using the centered log-ratio transform.</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co"># This is the input for all further analyses.</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>x <span class="ot">&lt;-</span> <span class="fu">aldex.clr</span>(<span class="fu">assay</span>(tse), tse<span class="sc">$</span>patient_status)     </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The t-test:</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/aldex2-ttest_04190e3cd56d73fe6ebec3b03004f9b8">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculates expected values of the Welch's t-test and Wilcoxon rank</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># test on the data returned by aldex.clr</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>x_tt <span class="ot">&lt;-</span> <span class="fu">aldex.ttest</span>(x, <span class="at">paired.test =</span> <span class="cn">FALSE</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Effect sizes:</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/aldex2-effect_ae102f082269047751da7a45cefd9d8f">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Determines the median clr abundance of the feature in all samples and in</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="co"># groups, the median difference between the two groups, the median variation</span></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="co"># within each group and the effect size, which is the median of the ratio</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a><span class="co"># of the between group difference and the larger of the variance within groups</span></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>x_effect <span class="ot">&lt;-</span> <span class="fu">aldex.effect</span>(x, <span class="at">CI =</span> <span class="cn">TRUE</span>, <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a><span class="co"># combine all outputs </span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>aldex_out <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(x_tt, x_effect)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now, we can create a so called Bland-Altman or MA plot (left). It shows the association between the relative abundance and the magnitude of the difference per sample. Next to that, we can also create a plot that shows the dispersion on the x-axis instead of log-ratio abundance. Red dots represent genera that are differentially abundant (<span class="math inline">\(q \leq 0.1\)</span>) between the 2 groups. Black points are rare taxa and grey ones are abundant taxa. The dashed line represent an effect size of 1. <span class="citation" data-cites="Gloor2016">@Gloor2016</span> provides more information on these plots.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/plot-aldex2_c038f5857e54fc1fa3da258b5d87d2d9">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mfrow =</span> <span class="fu">c</span>(<span class="dv">1</span>, <span class="dv">2</span>))</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">aldex.plot</span>(aldex_out,</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>           <span class="at">type =</span> <span class="st">"MA"</span>,</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>           <span class="at">test =</span> <span class="st">"welch"</span>,</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>           <span class="at">xlab =</span> <span class="st">"Log-ratio abundance"</span>,</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>           <span class="at">ylab =</span> <span class="st">"Difference"</span>,</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>           <span class="at">cutoff =</span> <span class="fl">0.05</span>)</span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a><span class="fu">aldex.plot</span>(aldex_out,</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>           <span class="at">type =</span> <span class="st">"MW"</span>,</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>           <span class="at">test =</span> <span class="st">"welch"</span>,</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>           <span class="at">xlab =</span> <span class="st">"Dispersion"</span>,</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>           <span class="at">ylab =</span> <span class="st">"Difference"</span>,</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>           <span class="at">cutoff =</span> <span class="fl">0.05</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="30_differential_abundance_files/figure-html/plot-aldex2-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>The evaluation as differential abundant in above plots is based on the corrected p-value. According to the ALDEx2 developers, the safest approach is to identify those features where the 95% CI of the effect size does not cross 0. As we can see in below table, this is not the case for any of the identified genera (see overlap column, which indicates the proportion of overlap). Also, the authors recommend to focus on effect sizes and CIs rather than interpreting the p-value. To keep the comparison simple, we will here use the p-value as decision criterion. But please be aware that the effect size together with the CI is a better answer to the question we are typically interested in.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/aldex2-res_65fd059c7c169adecdbff8e29e7fcad2">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>aldex_out <span class="sc">%&gt;%</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"Genus"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># here we choose the wilcoxon output rather than t-test output</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(wi.eBH <span class="sc">&lt;=</span> <span class="fl">0.05</span>)  <span class="sc">%&gt;%</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(Genus, we.eBH, wi.eBH, effect, overlap) <span class="sc">%&gt;%</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">Genus</th>
<th style="text-align: right;">we.eBH</th>
<th style="text-align: right;">wi.eBH</th>
<th style="text-align: right;">effect</th>
<th style="text-align: right;">overlap</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">[Ruminococcus]_gauvreauii_group</td>
<td style="text-align: right;">0.0664</td>
<td style="text-align: right;">0.0345</td>
<td style="text-align: right;">0.8184</td>
<td style="text-align: right;">0.1142</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="ancom-bc" class="level3" data-number="10.2.3">
<h3 data-number="10.2.3" class="anchored" data-anchor-id="ancom-bc"><span class="header-section-number">10.2.3</span> ANCOM-BC</h3>
<p>The analysis of composition of microbiomes with bias correction (ANCOM-BC) <span class="citation" data-cites="ancombc2020">[@ancombc2020]</span> is a recently developed method for differential abundance testing. It is based on an earlier published approach <span class="citation" data-cites="Mandal2015">[@Mandal2015]</span>. The previous version of ANCOM was among the methods that produced the most consistent results and is probably a conservative approach <span class="citation" data-cites="Nearing2022">[@Nearing2022]</span>. However, the new ANCOM-BC method operates quite differently compared to the former ANCOM method.</p>
<p>As the only method, ANCOM-BC incorporates the so called <em>sampling fraction</em> into the model. The latter term could be empirically estimated by the ratio of the library size to the microbial load. According to the authors, ignoring variations in this sampling fraction would bias DAA results. Furthermore, this method provides p-values and confidence intervals for each taxon. It also controls the FDR and it is computationally simple to implement.</p>
<p>Note that the original method was implemented in the <code>ancombc()</code> function (see <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/ANCOMBC/inst/doc/ANCOMBC.html">extended tutorial</a>). The method has since then been updated and new features have been added to enable multi-group comparisons and repeated measurements among other improvements. We do not cover the more advanced features of ANCOMBC in this tutorial as these features are documented in detail in this <a href="https://www.bioconductor.org/packages/release/bioc/vignettes/ANCOMBC/inst/doc/ANCOMBC2.html">tutorial</a>.</p>
<p>We now proceed with a simple example. First, we specify a formula. In this formula, other covariates could potentially be included to adjust for confounding. We show this further below. Again, please make sure to check the <a href="https://rdrr.io/github/FrederickHuangLin/ANCOMBC/man/ancombc.html">function documentation</a> as well as the linked tutorials to learn about the additional arguments that we specify.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/run-ancombc_35f307f985e458ea76e207b3d42fef19">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load package</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ANCOMBC)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Run ANCOM-BC at the genus level and only including the prevalent genera</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>ancombc2_out <span class="ot">&lt;-</span> <span class="fu">ancombc2</span>(<span class="at">data =</span> tse,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>                         <span class="at">assay_name =</span> <span class="st">"counts"</span>,</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>                         <span class="at">fix_formula =</span> <span class="st">"patient_status"</span>,</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>                         <span class="at">p_adj_method =</span> <span class="st">"fdr"</span>,</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>                         <span class="at">prv_cut =</span> <span class="dv">0</span>,</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>                         <span class="at">group =</span> <span class="st">"patient_status"</span>,</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>                         <span class="at">struc_zero =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>                         <span class="at">neg_lb =</span> <span class="cn">TRUE</span>,</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>                         <span class="co"># multi group comparison is deactivated automatically</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>                         <span class="at">global =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The object <code>out</code> contains all model output. Again, see the <a href="https://rdrr.io/github/FrederickHuangLin/ANCOMBC/man/ancombc.html">documentation of the function</a> under <strong>Value</strong> for details. Our question whether taxa are differentially abundant can be answered by looking at the <code>res</code> object, which contains dataframes with the coefficients, standard errors, p-values and q-values. Below we show the first entries of this dataframe.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/ancombc-res_3e9b1aedb8db25f043c1af2d3382d6a6">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># store the FDR adjusted results </span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>ancombc2_out<span class="sc">$</span>res <span class="sc">%&gt;%</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(taxon, lfc_patient_statusControl, q_patient_statusControl) <span class="sc">%&gt;%</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(q_patient_statusControl <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">%&gt;%</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(q_patient_statusControl) <span class="sc">%&gt;%</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>() <span class="sc">%&gt;%</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 39%">
<col style="width: 31%">
<col style="width: 29%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">taxon</th>
<th style="text-align: right;">lfc_patient_statusControl</th>
<th style="text-align: right;">q_patient_statusControl</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Subdoligranulum</td>
<td style="text-align: right;">1.909</td>
<td style="text-align: right;">0.0010</td>
</tr>
<tr class="even">
<td style="text-align: left;">Ruminococcus_1</td>
<td style="text-align: right;">2.915</td>
<td style="text-align: right;">0.0010</td>
</tr>
<tr class="odd">
<td style="text-align: left;">[Ruminococcus]_gauvreauii_group</td>
<td style="text-align: right;">1.520</td>
<td style="text-align: right;">0.0014</td>
</tr>
<tr class="even">
<td style="text-align: left;">[Eubacterium]_rectale_group</td>
<td style="text-align: right;">1.361</td>
<td style="text-align: right;">0.0050</td>
</tr>
<tr class="odd">
<td style="text-align: left;">[Clostridium]_innocuum_group</td>
<td style="text-align: right;">1.455</td>
<td style="text-align: right;">0.0052</td>
</tr>
<tr class="even">
<td style="text-align: left;">Dielma</td>
<td style="text-align: right;">1.166</td>
<td style="text-align: right;">0.0052</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="maaslin2" class="level3" data-number="10.2.4">
<h3 data-number="10.2.4" class="anchored" data-anchor-id="maaslin2"><span class="header-section-number">10.2.4</span> MaAsLin2</h3>
<p>Let us next illustrate MaAsLin2 <span class="citation" data-cites="Mallick2020">[@Mallick2020]</span>. This method is based on generalized linear models and flexible for different study designs and covariate structures. For details, check their <a href="https://github.com/biobakery/biobakery/wiki/maaslin2">Biobakery tutorial</a>.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/run-maaslin2_039ef67a63a66345ad4c3ab8b1e20df1">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load package</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Maaslin2)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a><span class="co"># maaslin expects features as columns and samples as rows </span></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a><span class="co"># for both the abundance table as well as metadata </span></span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a><span class="co"># We can specify different GLMs/normalizations/transforms.</span></span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a><span class="co"># specifying a ref is especially important if you have more than 2 levels</span></span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>maaslin2_out <span class="ot">&lt;-</span> <span class="fu">Maaslin2</span>(<span class="at">input_data =</span> <span class="fu">as.data.frame</span>(<span class="fu">t</span>(<span class="fu">assay</span>(tse))),</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>                         <span class="at">input_metadata =</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse)),</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>                         <span class="at">output =</span> <span class="st">"DAA example"</span>,</span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>                         <span class="at">transform =</span> <span class="st">"AST"</span>,</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>                         <span class="at">fixed_effects =</span> <span class="st">"patient_status"</span>,</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>                         <span class="co"># you can also fit MLM by specifying random effects</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>                         <span class="co"># random_effects = c(...),</span></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>                         <span class="at">reference =</span> <span class="st">"patient_status,Control"</span>,</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>                         <span class="at">normalization =</span> <span class="st">"TSS"</span>,</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>                         <span class="at">standardize =</span> <span class="cn">FALSE</span>,</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>                         <span class="co"># filtering was previously performed</span></span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a>                         <span class="at">min_prevalence =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Which genera are identified as differentially abundant? (leave out “head” to see all).</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/maaslin2-res_1b9c9d5aff6cd4f9fbda36823ec1ffca">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>maaslin2_out<span class="sc">$</span>results <span class="sc">%&gt;%</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(qval <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">%&gt;%</span></span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 28%">
<col style="width: 12%">
<col style="width: 5%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 6%">
<col style="width: 16%">
<col style="width: 6%">
<col style="width: 2%">
<col style="width: 9%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">feature</th>
<th style="text-align: left;">metadata</th>
<th style="text-align: left;">value</th>
<th style="text-align: right;">coef</th>
<th style="text-align: right;">stderr</th>
<th style="text-align: right;">pval</th>
<th style="text-align: left;">name</th>
<th style="text-align: right;">qval</th>
<th style="text-align: right;">N</th>
<th style="text-align: right;">N.not.zero</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">X.Ruminococcus._gauvreauii_group</td>
<td style="text-align: left;">patient_status</td>
<td style="text-align: left;">ADHD</td>
<td style="text-align: right;">-0.0674</td>
<td style="text-align: right;">0.0133</td>
<td style="text-align: right;">0.0000</td>
<td style="text-align: left;">patient_statusADHD</td>
<td style="text-align: right;">0.0015</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">21</td>
</tr>
<tr class="even">
<td style="text-align: left;">Faecalibacterium</td>
<td style="text-align: left;">patient_status</td>
<td style="text-align: left;">ADHD</td>
<td style="text-align: right;">0.1223</td>
<td style="text-align: right;">0.0372</td>
<td style="text-align: right;">0.0030</td>
<td style="text-align: left;">patient_statusADHD</td>
<td style="text-align: right;">0.0448</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">11</td>
</tr>
<tr class="odd">
<td style="text-align: left;">X.Clostridium._innocuum_group</td>
<td style="text-align: left;">patient_status</td>
<td style="text-align: left;">ADHD</td>
<td style="text-align: right;">-0.0692</td>
<td style="text-align: right;">0.0213</td>
<td style="text-align: right;">0.0033</td>
<td style="text-align: left;">patient_statusADHD</td>
<td style="text-align: right;">0.0448</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">25</td>
</tr>
<tr class="even">
<td style="text-align: left;">Catabacter</td>
<td style="text-align: left;">patient_status</td>
<td style="text-align: left;">ADHD</td>
<td style="text-align: right;">0.0295</td>
<td style="text-align: right;">0.0092</td>
<td style="text-align: right;">0.0037</td>
<td style="text-align: left;">patient_statusADHD</td>
<td style="text-align: right;">0.0448</td>
<td style="text-align: right;">27</td>
<td style="text-align: right;">9</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>This will create a folder that is called like in the output specified above. It contains also figures to visualize difference between significant taxa.</p>
</section>
<section id="linda" class="level3" data-number="10.2.5">
<h3 data-number="10.2.5" class="anchored" data-anchor-id="linda"><span class="header-section-number">10.2.5</span> LinDA</h3>
<p>Lastly, we cover linear models for differential abundance analysis of microbiome compositional data (<span class="citation" data-cites="Zhou2022">@Zhou2022</span>). This is very similar to ANCOMBC with few differences: 1) LinDA corrects for the compositional bias differently using the mode of all regression coefficients. 2) it is faster (100x-1000x than ANCOMBC and according to the authors); 3) it supports hierarchical models. The latest ANCOMBC versions are also supporting hierarchical models. Nevertheless, LinDA seems a promising tool that achieves a very good power/fdr trade-off together with ANCOMBC according to the review. The speed improvements might make it critical especially for datasets that have higher sample or feature set sizes.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/run-linda_e863c18ea835069c6dd41259fe38cab3">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load package</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MicrobiomeStat)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Run LinDA</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>linda_out <span class="ot">&lt;-</span> <span class="fu">linda</span>(<span class="at">feature.dat =</span> <span class="fu">as.data.frame</span>(<span class="fu">assay</span>(tse)),</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">meta.dat =</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse)),</span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true" tabindex="-1"></a>                   <span class="at">formula =</span> <span class="st">"~ patient_status"</span>,</span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true" tabindex="-1"></a>                   <span class="at">alpha =</span> <span class="fl">0.05</span>,</span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true" tabindex="-1"></a>                   <span class="at">prev.filter =</span> <span class="dv">0</span>,</span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true" tabindex="-1"></a>                   <span class="at">mean.abund.filter =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0  features are filtered!
The filtered data has  27  samples and  49  features will be tested!
Pseudo-count approach is used.
Fit linear models ...
Completed.</code></pre>
</div>
</div>
<div class="cell" data-hash="30_differential_abundance_cache/html/linda-res_ce3e796df8d5812791196c8b556a17b1">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># List genera for which H0 could be rejected:</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>linda_out<span class="sc">$</span>output<span class="sc">$</span>patient_statusControl <span class="sc">%&gt;%</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(reject) <span class="sc">%&gt;%</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(stat, padj) <span class="sc">%&gt;%</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"feature"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">feature</th>
<th style="text-align: right;">stat</th>
<th style="text-align: right;">padj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Faecalibacterium</td>
<td style="text-align: right;">-4.250</td>
<td style="text-align: right;">0.0092</td>
</tr>
<tr class="even">
<td style="text-align: left;">[Ruminococcus]_gauvreauii_group</td>
<td style="text-align: right;">4.110</td>
<td style="text-align: right;">0.0092</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Catabacter</td>
<td style="text-align: right;">-3.379</td>
<td style="text-align: right;">0.0390</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="zicoseq" class="level3" data-number="10.2.6">
<h3 data-number="10.2.6" class="anchored" data-anchor-id="zicoseq"><span class="header-section-number">10.2.6</span> ZicoSeq</h3>
<p>Subsequently, we demonstrate DAA with ZicoSeq, a method based on linear models and permutation. Further details can be found in <a href="https://cran.r-project.org/web/packages/GUniFrac/vignettes/ZicoSeq.html">this tutorial</a>. This approach has been assessed to exhibit high power and a low false discovery rate, which has the following components:</p>
<ol type="1">
<li>Winsorization to decrease the influence of outliers;</li>
<li>Posterior sampling based on a beta mixture prior to address sampling variability and zero inflation;</li>
<li>Reference-based multiple-stage normalization to address compositional effects;</li>
</ol>
<div class="cell" data-hash="30_differential_abundance_cache/html/run-zicoseq_0308e36eb04ab1dcd7412867643946a0">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load package</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(GUniFrac)</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a>zicoseq_out <span class="ot">&lt;-</span> <span class="fu">ZicoSeq</span>(<span class="at">feature.dat =</span> <span class="fu">as.matrix</span>(<span class="fu">assay</span>(tse)),</span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">meta.dat =</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse)),</span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a>                       <span class="at">grp.name =</span> <span class="st">"patient_status"</span>,</span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a>                       <span class="at">feature.dat.type =</span> <span class="st">"count"</span>,</span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a>                       <span class="at">return.feature.dat =</span> <span class="cn">TRUE</span>,</span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a>                       <span class="at">prev.filter =</span> <span class="dv">0</span>,</span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a>                       <span class="at">mean.abund.filter =</span> <span class="dv">0</span>,</span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>                       <span class="at">max.abund.filter =</span> <span class="dv">0</span>,</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>                       <span class="at">perm.no =</span> <span class="dv">999</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>For sample size less than 40, posterior sampling will not be used!
0  features are filtered!
The data has  27  samples and  49  features will be tested!
On average,  1  outlier counts will be replaced for each feature!
Finding the references ...
Permutation testing ...
...................................................................................................
...................................................................................................
...................................................................................................
...................................................................................................
...................................................................................................
...................................................................................................
Completed!</code></pre>
</div>
</div>
<div class="cell" data-hash="30_differential_abundance_cache/html/zicoseq-res_89b39ee979647ec62c4c66f0c875e579">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>zicoseq_res <span class="ot">&lt;-</span> <span class="fu">cbind.data.frame</span>(<span class="at">p.raw =</span> zicoseq_out<span class="sc">$</span>p.raw,</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>                                <span class="at">p.adj.fdr =</span> zicoseq_out<span class="sc">$</span>p.adj.fdr)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>zicoseq_res <span class="sc">%&gt;%</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(p.adj.fdr <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">%&gt;%</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(p.adj.fdr) <span class="sc">%&gt;%</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">p.raw</th>
<th style="text-align: right;">p.adj.fdr</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">[Ruminococcus]_gauvreauii_group</td>
<td style="text-align: right;">0.001</td>
<td style="text-align: right;">0.005</td>
</tr>
<tr class="even">
<td style="text-align: left;">Faecalibacterium</td>
<td style="text-align: right;">0.001</td>
<td style="text-align: right;">0.024</td>
</tr>
<tr class="odd">
<td style="text-align: left;">[Clostridium]_innocuum_group</td>
<td style="text-align: right;">0.003</td>
<td style="text-align: right;">0.024</td>
</tr>
<tr class="even">
<td style="text-align: left;">Catabacter</td>
<td style="text-align: right;">0.006</td>
<td style="text-align: right;">0.043</td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="cell" data-hash="30_differential_abundance_cache/html/plot-zicoseq_50cc109d0e8949415c9bf265a912ec57">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="do">## x-axis is the effect size: R2 * direction of coefficient</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ZicoSeq.plot</span>(<span class="at">ZicoSeq.obj =</span> zicoseq_out,</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>             <span class="at">pvalue.type =</span> <span class="st">'p.adj.fdr'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="30_differential_abundance_files/figure-html/plot-zicoseq-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="philr" class="level3" data-number="10.2.7">
<h3 data-number="10.2.7" class="anchored" data-anchor-id="philr"><span class="header-section-number">10.2.7</span> PhILR</h3>
<p>PhILR is a tree-based method that tests group-wise associations based on balances. A detailed introduction to this method is available in <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/philr/inst/doc/philr-intro.html">this Bioconductor tutorial</a>.</p>
</section>
<section id="comparison-of-methods" class="level3" data-number="10.2.8">
<h3 data-number="10.2.8" class="anchored" data-anchor-id="comparison-of-methods"><span class="header-section-number">10.2.8</span> Comparison of methods</h3>
<p>Although the methods described above yield unidentical results, they are expected to agree on a few differentially abundant taxa. To draw more informed conclusions, it is good practice to compare the outcomes of different methods in terms of found features, their effect sizes and significances, as well as other method-specific aspects. Such comparative approach is outlined in <a href="98_exercises.html#compare-daa-methods">this exercise</a>.</p>
</section>
</section>
<section id="daa-with-confounding" class="level2" data-number="10.3">
<h2 data-number="10.3" class="anchored" data-anchor-id="daa-with-confounding"><span class="header-section-number">10.3</span> DAA with confounding</h2>
<p>Confounders can be defined as variables that are related to and affect the apparent dynamics between the response and the main independent variable. They are common in experimental studies. Generally, they can be classified into 3 groups:</p>
<ul>
<li><p>Biological confounders, such as age and sex</p></li>
<li><p>Technical confounders produced during sample collection, processing and analysis</p></li>
<li><p>Confounders resulting from experimental models, such as batch effects and sample history</p></li>
</ul>
<p>Controlling for confounders is an important practice to reach an unbiased conclusion. To perform causal inference, it is crucial that the method is able to include confounders in the model. This is not possible with statistical tests of general use, such as the Wilcoxon test. In contrast, methods that target DAA, such as those described in this chapter, allow controlling for confounders. In the following examples, we will perform DAA with a main independent variable and a few confounders.</p>
<section id="selecting-confounders" class="level3" data-number="10.3.1">
<h3 data-number="10.3.1" class="anchored" data-anchor-id="selecting-confounders"><span class="header-section-number">10.3.1</span> Selecting confounders</h3>
<p>In addition to patient status, we will now control for two confounders: cohort and library size. The former is a categorical variable with three factors, whereas the latter is a discrete numerical variable. Remarkably, most DAA methods accept these two and several other data types.</p>
<p>For demonstration, library size is treated as a confounder and included in the formulas of the DAA methods. Although this is a satisfactory approach to control for uneven sequencing efforts across samples, rarefaction generally represents a better solution <span class="citation" data-cites="Schloss2023">[@Schloss2023]</span>. With that said, library size can be readily computed and added to the colData.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/daa-data-libsize_f4c10c3a0396c19f1bd6ebb069e59a8d">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Compute and store library size in colData</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="fu">colData</span>(tse)<span class="sc">$</span>library_size <span class="ot">&lt;-</span> <span class="fu">colSums</span>(<span class="fu">assay</span>(tse, <span class="st">"counts"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="ancom-bc-1" class="level3" data-number="10.3.2">
<h3 data-number="10.3.2" class="anchored" data-anchor-id="ancom-bc-1"><span class="header-section-number">10.3.2</span> ANCOM-BC</h3>
<p>Here, confounders can be added to the formula along with patient status, the main outcome variable. This way, the model evaluates whether differentially abundant taxa are associated with one of the variables when the other two are kept constant.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/run-adj-ancombc_333b3168fb8224b75647d2cfcd1726cf">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># perform the analysis </span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>ancombc2_out <span class="ot">&lt;-</span> <span class="fu">ancombc2</span>(tse,</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>                         <span class="at">assay_name =</span> <span class="st">"counts"</span>,</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true" tabindex="-1"></a>                         <span class="at">fix_formula =</span> <span class="st">"patient_status + cohort + library_size"</span>,</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true" tabindex="-1"></a>                         <span class="at">p_adj_method =</span> <span class="st">"fdr"</span>,</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true" tabindex="-1"></a>                         <span class="at">lib_cut =</span> <span class="dv">0</span>,</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true" tabindex="-1"></a>                         <span class="at">group =</span> <span class="st">"patient_status"</span>, </span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true" tabindex="-1"></a>                         <span class="at">struc_zero =</span> <span class="cn">TRUE</span>, </span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true" tabindex="-1"></a>                         <span class="at">neg_lb =</span> <span class="cn">TRUE</span>,</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true" tabindex="-1"></a>                         <span class="at">alpha =</span> <span class="fl">0.05</span>,</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true" tabindex="-1"></a>                         <span class="co"># multi-group comparison is deactivated automatically</span></span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true" tabindex="-1"></a>                         <span class="at">global =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>In the output, each taxon is assigned with several effect sizes (lfc, which stands for log-fold change) and adjusted p-values (q). For categorical variables such as patient status and cohort, the statistics indicate whether the abundance of a given taxon is significantly different between the specified group (column name) and the reference group (the group that does not appear in the column names), whereas for numerical variables such as library size, they indicate whether the abundance of a given taxon varies with that variable.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/adj-ancombc-res_95b8124d5ee2f40c626c9cfd22241049">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a>ancombc2_out<span class="sc">$</span>res <span class="sc">%&gt;%</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="fu">starts_with</span>(<span class="fu">c</span>(<span class="st">"taxon"</span>, <span class="st">"lfc"</span>, <span class="st">"q"</span>))) <span class="sc">%&gt;%</span></span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(q_patient_statusControl) <span class="sc">%&gt;%</span></span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>() <span class="sc">%&gt;%</span></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 13%">
<col style="width: 7%">
<col style="width: 12%">
<col style="width: 8%">
<col style="width: 8%">
<col style="width: 7%">
<col style="width: 6%">
<col style="width: 11%">
<col style="width: 7%">
<col style="width: 7%">
<col style="width: 7%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">taxon</th>
<th style="text-align: right;">lfc_(Intercept)</th>
<th style="text-align: right;">lfc_patient_statusControl</th>
<th style="text-align: right;">lfc_cohortCohort_2</th>
<th style="text-align: right;">lfc_cohortCohort_3</th>
<th style="text-align: right;">lfc_library_size</th>
<th style="text-align: right;">q_(Intercept)</th>
<th style="text-align: right;">q_patient_statusControl</th>
<th style="text-align: right;">q_cohortCohort_2</th>
<th style="text-align: right;">q_cohortCohort_3</th>
<th style="text-align: right;">q_library_size</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Akkermansia</td>
<td style="text-align: right;">-0.8599</td>
<td style="text-align: right;">0.2309</td>
<td style="text-align: right;">0.2876</td>
<td style="text-align: right;">0.4635</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.9147</td>
</tr>
<tr class="even">
<td style="text-align: left;">Hungatella</td>
<td style="text-align: right;">-0.0695</td>
<td style="text-align: right;">-0.3270</td>
<td style="text-align: right;">-0.1397</td>
<td style="text-align: right;">-0.1151</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0593</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Ruminococcaceae_UCG-013</td>
<td style="text-align: right;">-0.9344</td>
<td style="text-align: right;">-0.3371</td>
<td style="text-align: right;">0.6599</td>
<td style="text-align: right;">-0.0231</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0050</td>
</tr>
<tr class="even">
<td style="text-align: left;">Bacteroides</td>
<td style="text-align: right;">-0.3081</td>
<td style="text-align: right;">-0.7575</td>
<td style="text-align: right;">0.1482</td>
<td style="text-align: right;">0.7615</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.9650</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Escherichia-Shigella</td>
<td style="text-align: right;">-1.1600</td>
<td style="text-align: right;">-0.5157</td>
<td style="text-align: right;">1.3093</td>
<td style="text-align: right;">0.2340</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0397</td>
</tr>
<tr class="even">
<td style="text-align: left;">[Clostridium]_innocuum_group</td>
<td style="text-align: right;">-0.7590</td>
<td style="text-align: right;">0.7781</td>
<td style="text-align: right;">-0.1629</td>
<td style="text-align: right;">0.1688</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0</td>
<td style="text-align: right;">0.0593</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="linda-1" class="level3" data-number="10.3.3">
<h3 data-number="10.3.3" class="anchored" data-anchor-id="linda-1"><span class="header-section-number">10.3.3</span> LinDA</h3>
<p>As in the previous method, confounders can be included in the formula with the main outcome variable.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/run-adj-linda_18fdba30230cf1b1a660e056bf8362e5">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>linda_out <span class="ot">&lt;-</span> <span class="fu">linda</span>(<span class="fu">as.data.frame</span>(<span class="fu">assay</span>(tse, <span class="st">"counts"</span>)),</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>                   <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse)),</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>                   <span class="at">formula =</span> <span class="st">"~ patient_status + cohort + library_size"</span>,</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">alpha =</span> <span class="fl">0.05</span>,</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">prev.filter =</span> <span class="dv">0</span>,</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>                   <span class="at">mean.abund.filter =</span> <span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>0  features are filtered!
The filtered data has  27  samples and  49  features will be tested!
Imputation approach is used.
Fit linear models ...
Completed.</code></pre>
</div>
</div>
<p>The model returns an output for every variable included in the formula. Normally, only the results on the main outcome variable are relevant and can be retrieved as shown below. However, the statistics on the confounders can be similarly obtained by accessing the corresponding items from the output object.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/adj-linda-res_6107638ef2568f422d7d362bf2f90561">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Select results for the patient status</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>linda_res <span class="ot">&lt;-</span> linda_out<span class="sc">$</span>output<span class="sc">$</span>patient_statusControl</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a>linda_res <span class="sc">%&gt;%</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(reject) <span class="sc">%&gt;%</span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(log2FoldChange, stat, padj) <span class="sc">%&gt;%</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var =</span> <span class="st">"feature"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>() <span class="sc">%&gt;%</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;">feature</th>
<th style="text-align: right;">log2FoldChange</th>
<th style="text-align: right;">stat</th>
<th style="text-align: right;">padj</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Faecalibacterium</td>
<td style="text-align: right;">-5.921</td>
<td style="text-align: right;">-4.521</td>
<td style="text-align: right;">0.0041</td>
</tr>
<tr class="even">
<td style="text-align: left;">Erysipelatoclostridium</td>
<td style="text-align: right;">3.743</td>
<td style="text-align: right;">3.010</td>
<td style="text-align: right;">0.0451</td>
</tr>
<tr class="odd">
<td style="text-align: left;">[Ruminococcus]_gauvreauii_group</td>
<td style="text-align: right;">4.240</td>
<td style="text-align: right;">4.534</td>
<td style="text-align: right;">0.0041</td>
</tr>
<tr class="even">
<td style="text-align: left;">Barnesiella</td>
<td style="text-align: right;">-3.878</td>
<td style="text-align: right;">-3.116</td>
<td style="text-align: right;">0.0411</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Ruminococcaceae_UCG-014</td>
<td style="text-align: right;">-3.062</td>
<td style="text-align: right;">-3.637</td>
<td style="text-align: right;">0.0193</td>
</tr>
<tr class="even">
<td style="text-align: left;">Butyricicoccus</td>
<td style="text-align: right;">-2.357</td>
<td style="text-align: right;">-3.127</td>
<td style="text-align: right;">0.0411</td>
</tr>
</tbody>
</table>
</div>
</div>
<p>The output shows effect sizes in terms of log-fold changes and a derived statistic (stat) as well as the corresponding adjusted p-values for differences in abundance of each taxon between the control and treated group.</p>
</section>
<section id="zicoseq-1" class="level3" data-number="10.3.4">
<h3 data-number="10.3.4" class="anchored" data-anchor-id="zicoseq-1"><span class="header-section-number">10.3.4</span> ZicoSeq</h3>
<p>For this method, confounders can be added as a list to the <code>adj.name</code> argument.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/run-adj-zicoseq_b1fae05b7dc57704370883639e071f8d">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123</span>)</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>zicoseq_out <span class="ot">&lt;-</span> <span class="fu">ZicoSeq</span>(<span class="at">feature.dat =</span> <span class="fu">as.matrix</span>(<span class="fu">assay</span>(tse)),</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">meta.dat =</span> <span class="fu">as.data.frame</span>(<span class="fu">colData</span>(tse)),</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">grp.name =</span> <span class="st">"patient_status"</span>,</span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>                       <span class="at">adj.name =</span> <span class="fu">c</span>(<span class="st">"cohort"</span>, <span class="st">"library_size"</span>), </span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>                       <span class="at">feature.dat.type =</span> <span class="st">"count"</span>,</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>                       <span class="at">return.feature.dat =</span> <span class="cn">TRUE</span>,</span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>                       <span class="at">prev.filter =</span> <span class="dv">0</span>,</span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>                       <span class="at">mean.abund.filter =</span> <span class="dv">0</span>,</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>                       <span class="at">max.abund.filter =</span> <span class="dv">0</span>,</span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>                       <span class="at">perm.no =</span> <span class="dv">999</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>For sample size less than 40, posterior sampling will not be used!
0  features are filtered!
The data has  27  samples and  49  features will be tested!
On average,  1  outlier counts will be replaced for each feature!
Finding the references ...
Permutation testing ...
...................................................................................................
...................................................................................................
...................................................................................................
...................................................................................................
...................................................................................................
...................................................................................................
Completed!</code></pre>
</div>
</div>
<p>The output shows the raw and adjusted p-values for clinical status.</p>
<div class="cell" data-hash="30_differential_abundance_cache/html/adj-zicoseq-res_b39eab67c1834120d2e32d525a5a8f6b">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>zicoseq_res <span class="ot">&lt;-</span> <span class="fu">cbind.data.frame</span>(<span class="at">p.raw =</span> zicoseq_out<span class="sc">$</span>p.raw,</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>                                <span class="at">p.adj.fdr =</span> zicoseq_out<span class="sc">$</span>p.adj.fdr)</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>zicoseq_res <span class="sc">%&gt;%</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(p.adj.fdr <span class="sc">&lt;</span> <span class="fl">0.05</span>) <span class="sc">%&gt;%</span></span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>() <span class="sc">%&gt;%</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>  knitr<span class="sc">::</span><span class="fu">kable</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th style="text-align: left;"></th>
<th style="text-align: right;">p.raw</th>
<th style="text-align: right;">p.adj.fdr</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Faecalibacterium</td>
<td style="text-align: right;">0.002</td>
<td style="text-align: right;">0.0224</td>
</tr>
<tr class="even">
<td style="text-align: left;">[Clostridium]_innocuum_group</td>
<td style="text-align: right;">0.003</td>
<td style="text-align: right;">0.0238</td>
</tr>
<tr class="odd">
<td style="text-align: left;">[Ruminococcus]_gauvreauii_group</td>
<td style="text-align: right;">0.001</td>
<td style="text-align: right;">0.0005</td>
</tr>
<tr class="even">
<td style="text-align: left;">Ruminococcus_2</td>
<td style="text-align: right;">0.002</td>
<td style="text-align: right;">0.0300</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Ruminococcaceae_UCG-014</td>
<td style="text-align: right;">0.004</td>
<td style="text-align: right;">0.0369</td>
</tr>
<tr class="even">
<td style="text-align: left;">Catabacter</td>
<td style="text-align: right;">0.003</td>
<td style="text-align: right;">0.0224</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
</section>
<section id="additional-resources" class="level2" data-number="10.4">
<h2 data-number="10.4" class="anchored" data-anchor-id="additional-resources"><span class="header-section-number">10.4</span> Additional resources</h2>
<p>DAA can be performed by several means. Although most of them provide similar functionality, some may be more suitable than others given a certain study design or data type. Commonly used DAA tools include:</p>
<ul>
<li>ALDEx2 <span class="citation" data-cites="Gloor2016">[@Gloor2016]</span></li>
<li>ANCOM <span class="citation" data-cites="Mandal2015">[@Mandal2015]</span></li>
<li>ANCOMBC <span class="citation" data-cites="ancombc2020">[@ancombc2020]</span></li>
<li>corncob <span class="citation" data-cites="Martin2021">[@Martin2021]</span></li>
<li>DACOMP <span class="citation" data-cites="Brill2019">[@Brill2019]</span></li>
<li>DESeq2 <span class="citation" data-cites="Love2014">[@Love2014]</span></li>
<li>eBay <span class="citation" data-cites="Liu2020">[@Liu2020]</span></li>
<li>edgeR <span class="citation" data-cites="Chen2016">[@Chen2016]</span></li>
<li>fastANCOM <span class="citation" data-cites="fastANCOM2022">[@fastANCOM2022]</span></li>
<li>LDM <span class="citation" data-cites="Hu2020">[@Hu2020]</span></li>
<li>lefser <span class="citation" data-cites="Khlebrodova2021">[@Khlebrodova2021]</span></li>
<li>limma <span class="citation" data-cites="Ritchie2015">[@Ritchie2015]</span></li>
<li>LinDA <span class="citation" data-cites="Zhou2022">[@Zhou2022]</span></li>
<li>MaAsLin2 <span class="citation" data-cites="Mallick2020">[@Mallick2020]</span></li>
<li>metagenomeSeq <span class="citation" data-cites="Paulson2017">[@Paulson2017]</span></li>
<li>Omnibus <span class="citation" data-cites="Omnibus2018">[@Omnibus2018]</span></li>
<li>RAIDA <span class="citation" data-cites="Sohn2015">[@Sohn2015]</span></li>
<li>t-test</li>
<li>Wilcoxon test</li>
<li>ZicoSeq <span class="citation" data-cites="Yang2022">[@Yang2022]</span></li>
<li>ZINQ <span class="citation" data-cites="Ling2021">[@Ling2021]</span></li>
</ul>


</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const disableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'prefetch';
    }
  }
  const enableStylesheet = (stylesheets) => {
    for (let i=0; i < stylesheets.length; i++) {
      const stylesheet = stylesheets[i];
      stylesheet.rel = 'stylesheet';
    }
  }
  const manageTransitions = (selector, allowTransitions) => {
    const els = window.document.querySelectorAll(selector);
    for (let i=0; i < els.length; i++) {
      const el = els[i];
      if (allowTransitions) {
        el.classList.remove('notransition');
      } else {
        el.classList.add('notransition');
      }
    }
  }
  const toggleColorMode = (alternate) => {
    // Switch the stylesheets
    const alternateStylesheets = window.document.querySelectorAll('link.quarto-color-scheme.quarto-color-alternate');
    manageTransitions('#quarto-margin-sidebar .nav-link', false);
    if (alternate) {
      enableStylesheet(alternateStylesheets);
      for (const sheetNode of alternateStylesheets) {
        if (sheetNode.id === "quarto-bootstrap") {
          toggleBodyColorMode(sheetNode);
        }
      }
    } else {
      disableStylesheet(alternateStylesheets);
      toggleBodyColorPrimary();
    }
    manageTransitions('#quarto-margin-sidebar .nav-link', true);
    // Switch the toggles
    const toggles = window.document.querySelectorAll('.quarto-color-scheme-toggle');
    for (let i=0; i < toggles.length; i++) {
      const toggle = toggles[i];
      if (toggle) {
        if (alternate) {
          toggle.classList.add("alternate");     
        } else {
          toggle.classList.remove("alternate");
        }
      }
    }
    // Hack to workaround the fact that safari doesn't
    // properly recolor the scrollbar when toggling (#1455)
    if (navigator.userAgent.indexOf('Safari') > 0 && navigator.userAgent.indexOf('Chrome') == -1) {
      manageTransitions("body", false);
      window.scrollTo(0, 1);
      setTimeout(() => {
        window.scrollTo(0, 0);
        manageTransitions("body", true);
      }, 40);  
    }
  }
  const isFileUrl = () => { 
    return window.location.protocol === 'file:';
  }
  const hasAlternateSentinel = () => {  
    let styleSentinel = getColorSchemeSentinel();
    if (styleSentinel !== null) {
      return styleSentinel === "alternate";
    } else {
      return false;
    }
  }
  const setStyleSentinel = (alternate) => {
    const value = alternate ? "alternate" : "default";
    if (!isFileUrl()) {
      window.localStorage.setItem("quarto-color-scheme", value);
    } else {
      localAlternateSentinel = value;
    }
  }
  const getColorSchemeSentinel = () => {
    if (!isFileUrl()) {
      const storageValue = window.localStorage.getItem("quarto-color-scheme");
      return storageValue != null ? storageValue : localAlternateSentinel;
    } else {
      return localAlternateSentinel;
    }
  }
  let localAlternateSentinel = 'default';
  // Dark / light mode switch
  window.quartoToggleColorScheme = () => {
    // Read the current dark / light value 
    let toAlternate = !hasAlternateSentinel();
    toggleColorMode(toAlternate);
    setStyleSentinel(toAlternate);
  };
  // Ensure there is a toggle, if there isn't float one in the top right
  if (window.document.querySelector('.quarto-color-scheme-toggle') === null) {
    const a = window.document.createElement('a');
    a.classList.add('top-right');
    a.classList.add('quarto-color-scheme-toggle');
    a.href = "";
    a.onclick = function() { try { window.quartoToggleColorScheme(); } catch {} return false; };
    const i = window.document.createElement("i");
    i.classList.add('bi');
    a.appendChild(i);
    window.document.body.appendChild(a);
  }
  // Switch to dark mode if need be
  if (hasAlternateSentinel()) {
    toggleColorMode(true);
  } else {
    toggleColorMode(false);
  }
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
<nav class="page-navigation">
  <div class="nav-page nav-page-previous">
      <a href="./21_microbiome_community.html" class="pagination-link">
        <i class="bi bi-arrow-left-short"></i> <span class="nav-page-text"><span class="chapter-number">9</span>&nbsp; <span class="chapter-title">Community Composition</span></span>
      </a>          
  </div>
  <div class="nav-page nav-page-next">
      <a href="./40_machine_learning.html" class="pagination-link">
        <span class="nav-page-text"><span class="chapter-number">11</span>&nbsp; <span class="chapter-title">Machine Learning</span></span> <i class="bi bi-arrow-right-short"></i>
      </a>
  </div>
</nav>
</div> <!-- /content -->



</body></html>